---
title: "06-normalisation_and_dge"
output: html_document
date: "2023-06-20"
---

In this notebook we'll be looking at the gene-level count data from the RNA-seq.
We'll perform DGE tests for each bioproject separately, look at combining DGE related results using p-value metaanalysis, and look at normalising counts for downstream co-expression analysis.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


The first think is to load the data and see what we're dealing with.

```{r}
library(tidyverse)
# library(ruv)
library(DESeq2) # bioconductor
library(vsn) # bioconductor
library(edgeR) # bioconductor
library(ggplot2)
library(ggrepel)
library(ggExtra)
library(scales)
library(HTSFilter) # bioconductor
library(sva) # bioconductor
library("UpSetR")

library(ComplexHeatmap) # bioconductor
```

BiocManager::install("sva")
BiocManager::install("HTSFilter")
BiocManager::install("ComplexHeatmap")


```{r}
ggplot2::theme_set(ggplot2::theme_bw() +
  ggplot2::theme(
    strip.background = element_rect(
     color="black", fill="white", linewidth=0, linetype="solid"
    ),
    rect = element_rect(
      colour = "black",
      linewidth = 1,
      linetype = "solid"
    )
  )
)

okabe <- c('#000000', '#E69F00', '#56B4E9', '#009E73', '#F0E042', '#0072B2', '#D55E00', '#CC79A7', '#999999')
options(ggplot2.discrete.fill = list(okabe[seq(2, 9, 2)], okabe[2:8], okabe, rep(okabe, 10)))
options(ggplot2.discrete.colour = list(okabe[seq(2, 9, 2)], okabe[2:8], okabe, rep(okabe, 10)))

gg_shape <- function(gg, vals) {gg + scale_shape_manual(values = rep(c(15, 17:20), 100))}
```


```{r}
counts_raw <- readr::read_delim("output/feature_counts.tsv", delim = "\t") %>%
  mutate(Geneid = str_remove(Geneid, "gene-"))
head(counts_raw)
```

```{r}
counts <- counts_raw %>%
  mutate(id=Geneid) %>%
  select(-Chr, -Start, -End, -Strand, -Length, -Geneid)

# There must be a more elegant way to do this...
counts <- counts[, c("id", colnames(counts)[colnames(counts) != "id"])]
head(counts)
```


```{r}
meta <- readr::read_delim("input/sra_rnaseq2.tsv", delim = "\t", na = "-") %>%
  mutate(
    host_ = paste0(host_tidyname, host_accession_tidyname),
    organism_ = paste0(organism_tidyname, accession_tidyname),
  ) %>%
  filter(organism != "Sclerotinia trifoliorum")

head(meta)
```

```{r}
virus_mapper <- c("DT8SsHADV" = "SsHADV", "DT8" = "VF", "Z113OXSsNSRV1" = "SsNSRV1", "Z11OXSsNSRV1" = "SsNSRV1", "AH98" = "SsNSRV1", "1980" = "VF", "DK3SlaGemV1" = "SlaGemV1", "DK3" = "VF")

meta <- meta %>% mutate(
  infection_stage = case_when(
    (host_tidyname == "Psat") & (hpi < 24) ~ "E",
    (host_tidyname == "Psat") & (24 <= hpi) & (hpi < 48) ~ "M",
    (host_tidyname == "Psat") & (48 <= hpi) ~ "L",
    (host_tidyname == "Gmax") & (hpi < 48) ~ "E",
    (host_tidyname == "Gmax") & (48 <= hpi) & (hpi < 96) ~ "M",
    (host_tidyname == "Gmax") & (96 <= hpi) ~ "L",
    (host_tidyname == "Bnap") & (hpi < 12) ~ "E",
    (host_tidyname == "Bnap") & (12 <= hpi) & (hpi < 48) ~ "M",
    (host_tidyname == "Bnap") & (48 <= hpi) ~ "L",
    (host_tidyname == "Cari") & (hpi < 24) ~ "E",
    (host_tidyname == "Cari") & (24 <= hpi) & (hpi < 24) ~ "M",
    (host_tidyname == "Cari") & (72 <= hpi) ~ "L",
    TRUE ~ NA
  ),
  resistance = case_when(
    (host_tidyname == "Psat") & (host_accession == "Lifter") ~ "Susc",
    (host_tidyname == "Psat") & (host_accession == "PI240515") ~ "Res",
    (host_tidyname == "Gmax") & (host_accession_tidyname == "9144") ~ "Susc",
    (host_tidyname == "Gmax") & (host_accession_tidyname == "91145") ~ "Res",
    (host_tidyname == "Gmax") & (host_accession_tidyname == "ACColibri") ~ "Susc",
    (host_tidyname == "Gmax") & (host_accession_tidyname == "ACColibriOA") ~ "Res",
    (host_tidyname == "Cari") & (host_accession_tidyname == "Kyabra") ~ "Susc",
    (host_tidyname == "Cari") & (host_accession_tidyname == "HatTrick") ~ "Res",
    (host_tidyname == "Bnap") & (host_accession_tidyname == "NingRS1") ~ "Res",
    (host_tidyname == "Bnap") & (host_accession_tidyname == "1703") ~ "Res",
    (host_tidyname == "Lsat") & (host_accession_tidyname %in% c("RedGranoble", "58", "Pallone", "OakLeaf", "KaiserSelbstschluss", "Kahu", "CobhamGreen", "BloodyWarrior", "Ambassador", "Adriatica2", "68", "Simpson", "Saladin", "RomainDeBenicardo")) ~ "Susc",
    (host_tidyname == "Lsat") & (host_accession_tidyname %in% c("Prazan", "Iceberg", "BataviaTezier", "Aspen")) ~ "Res",
    TRUE ~ NA  
  ),
  virus = ifelse(accession_tidyname %in% names(virus_mapper), virus_mapper[accession_tidyname], NA)
) %>% mutate(
  infection_stage = factor(ifelse((!is.na(resistance)) & (!is.na(infection_stage)) & (resistance == "Res"), paste0(infection_stage, "Res"), infection_stage))
) %>% mutate(
  sample = factor(sample),
  host_tidyname = factor(host_tidyname),
  infection_stage = factor(infection_stage),
  resistance = factor(resistance),
  virus = factor(virus),
  hpi = factor(hpi),
  ipvsivtp = factor(ifelse(host_tidyname == "IV", "IV", ifelse(is.na(infection_stage), NA, paste0(host_tidyname, "_", infection_stage))))
)

head(meta %>% select(sra, bioproject, host_tidyname, sample, infection_stage, resistance, virus, ipvsivtp))
```


OK. The data is in, let's look at the counts.

```{r, fig.height=8, fig.width=14}
ggplot(meta, aes(x=sra, y=n_fragments_aligned, fill=included)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(bioproject), scales="free")
```

So you should see that there are several bioprojects that I've excluded on the basis of multimapping reads, generally low read coverage, etc.
The samples labelled "network only" are excluded from DGE because they are unreplicated. However they may be of interest to identify co-expressed genes.

You should also see that we have a pretty wide range of coverage.
Typically the most dramatic differences are between samples in planta and in vitro.

Sometimes a biological treatment will be split across multiple sequencing runs.
I'll check for this now.

```{r}
multi_biosample <- meta %>% group_by(biosample) %>% summarise(n = n_distinct(sra)) %>% filter(n > 1) %>% .[["biosample"]]

for (m in multi_biosample) {
  print(meta[meta$biosample == m, ])
}
```

Ok. So these all seem to be treatment replicates. I'm not going to worry about combining any counts.

We'll exclude any samples that we aren't going to keep right now.
And I want to convert the table to a long format which i can use to select subsets very easily.

```{r}
meta <- meta[meta$included %in% c("TRUE", "NETWORK_ONLY"), ]

lcounts <- dplyr::inner_join(
  meta,
  tidyr::pivot_longer(counts, -id, names_to = "sra", values_to = "count"),
  by = "sra"
) %>% mutate(cpm = count / (n_fragments_aligned / 1000000)) %>%
  select(-platform, -read_length, -strategy, -stranded, -n_fragments, -n_fragments_filtered, -n_fragments_aligned, -n_fragments_aligned_single, -n_fragments_aligned_multi, -notes, -DOI, -biosample)

head(lcounts)
```

Let's have a look at some PCA plots.

```{r}
tmp <- lcounts %>% tidyr::pivot_wider(id_cols = c(id), names_from = sra, values_from = cpm) %>%
  column_to_rownames("id") %>%
  as.matrix()

pca_raw <- princomp(tmp, fix_sign = TRUE)
rm(tmp)

prop_var <- ((pca_raw$sdev[1:10] / sum(pca_raw$sdev[1:10])) * 100) %>%
  data.frame(variance_explained = .) %>%
  rownames_to_column("PC") %>%
  mutate(PC = paste0("PC", sprintf("%02d", 1:n())))

ggplot(prop_var, aes(x=PC, y=variance_explained)) + geom_bar(stat = "identity")
```

The first PC explains a fairly large amount of variance.
I wasn't expecting to get much reduction in this dataset.

```{r}
pca <- as.data.frame(pca_raw$loadings[, 1:10])
colnames(pca) <- paste0("PC", sprintf("%02d", seq_len(ncol(pca))))
pca <- pca %>% rownames_to_column("sra")
pca <- right_join(meta, pca, by = "sra")
head(pca)
```


```{r, fig.width = 4.75, fig.height = 3.5}
gg <- ggplot(pca, aes(x=PC01, y=PC02, color=host_tidyname, shape=host_tidyname, fill=host_tidyname)) +
  geom_point(alpha = 0.7, size = 3)

gg <- gg_shape(gg, pca$host_tidyname)
print(gg)

gg <- ggplot(pca, aes(x=PC03, y=PC04, color=host_tidyname, fill=host_tidyname, shape=host_tidyname)) +
  geom_point(alpha = 0.7, size = 3)

gg <- gg_shape(gg, pca$host_tidyname)
gg
```

Well, the first and second axes seem to be picking something up with B. napus samples, and maybe some unusual in-vitro samples.
It looks like the B. napus samples are going to be dominating the data.
Possibly just because there's more of it.

```{r, fig.width = 6.25, fig.height = 3.5}
gg <- ggplot(pca, aes(x=PC01, y=PC02, color=bioproject, fill=bioproject, shape=bioproject)) +
  geom_point(alpha = 0.7, size = 3)

gg <- gg_shape(gg, pca$bioproject)
print(gg)

gg <- ggplot(pca, aes(x=PC03, y=PC04, color=bioproject, fill=bioproject, shape=bioproject)) +
  geom_point(alpha = 0.7, size = 3)

gg <- gg_shape(gg, pca$bioproject)
print(gg)
```

It does seem that the little clusters are at least coming from the same bioprojects, which is encouraging.
Maybe a heatmap would be nice to show how the different experiments are (dis)similar.


```{r}
# Utility function from https://jokergoo.github.io/2020/05/11/set-cell-width/height-in-the-heatmap/
calc_ht_size = function(ht, unit = "inches") {
    pdf(NULL)
    ht = draw(ht)
    w = ComplexHeatmap:::width(ht)
    w = convertX(w, unit, valueOnly = TRUE)
    h = ComplexHeatmap:::height(ht)
    h = convertY(h, unit, valueOnly = TRUE)
    dev.off()

    c(w, h)
}
```

We'll look at clustering the samples to see if natural groups fall out from a correlation matrix.

```{r, fig.width=12.03533, fig.height=8.11880}
tmp <- lcounts %>% tidyr::pivot_wider(id_cols = c(id), names_from = sra, values_from = cpm) %>%
  column_to_rownames("id") %>%
  as.matrix()

tmp <- log(tmp + 1, base = 2)

tmp_cor <- cor(tmp)
tmp_dist <- as.dist(1 - tmp_cor)
hcl <- hclust(tmp_dist, method = "average")

srrs <- colnames(tmp)[hcl$order]
ha <- HeatmapAnnotation(
  bioproject = meta %>% column_to_rownames("sra") %>% .[srrs, "bioproject"],
  host = meta %>% column_to_rownames("sra") %>% .[srrs, "host"]
)

ht <- Heatmap(
  tmp_cor,
  name = "samples",
  cluster_rows = hcl,
  cluster_columns = hcl,
  show_column_dend = FALSE,
  show_column_names = FALSE,
  show_row_names = FALSE,
  top_annotation = ha,
  height=nrow(tmp_cor) * unit(0.5, "mm"),
  width=ncol(tmp_cor) * unit(0.5, "mm")
)

rm(tmp_cor)
rm(tmp)

htsize <- calc_ht_size(ht)
htsize

pdf("output/06-samples_correlation_clustered.pdf", width = htsize[1], height = htsize[2])
ht
dev.off()

draw(ht)
```


So you can see from the top-colour bar that the clusters are a mixture of bioprojects and hosts.
Host tends to form larger blocks, but not all samples from the same host are positioned together.

Next we'll look at an unclustered matrix, so samples from the same bioproject should occur together.

```{r, fig.width=11.62194, fig.height=8.11880}
tmp <- lcounts %>% tidyr::pivot_wider(id_cols = c(id), names_from = sra, values_from = cpm) %>%
  column_to_rownames("id") %>%
  as.matrix()

tmp <- log(tmp + 1, base = 2)

tmp_cor <- cor(tmp)
#tmp_dist <- as.dist(1 - tmp_cor)
#hcl <- hclust(tmp_dist, method = "average")

#srrs <- colnames(tmp)[hcl$order]
ha <- HeatmapAnnotation(
  bioproject = meta %>% column_to_rownames("sra") %>% .[colnames(tmp), "bioproject"],
  host = meta %>% column_to_rownames("sra") %>% .[colnames(tmp), "host"]
)

ht <- Heatmap(
  tmp_cor,
  name = "samples",
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  show_column_dend = FALSE,
  show_column_names = FALSE,
  show_row_names = FALSE,
  top_annotation = ha,
  height=nrow(tmp_cor) * unit(0.5, "mm"),
  width=ncol(tmp_cor) * unit(0.5, "mm")
)

rm(tmp_cor)
rm(tmp)

htsize <- calc_ht_size(ht)
htsize

pdf("output/06-samples_correlation_bioproject.pdf", width = htsize[1], height = htsize[2])
ht
dev.off()

draw(ht)
```


And finally we'll look at host.


```{r, fig.width=11.62194, fig.height=8.11880}
tmp <- lcounts %>% tidyr::pivot_wider(id_cols = c(id), names_from = sra, values_from = cpm) %>%
  column_to_rownames("id") %>%
  as.matrix()

tmp <- log(tmp + 1, base = 2)
meta_tmp <- meta[order(meta$host, meta$host_accession, meta$hpi, meta$tissue, meta$accession, meta$plant_tissue, meta$treatment),]
tmp <- tmp[, meta_tmp$sra[meta_tmp$sra %in% colnames(tmp)]]

tmp_cor <- cor(tmp)
#tmp_dist <- as.dist(1 - tmp_cor)
#hcl <- hclust(tmp_dist, method = "average")

#srrs <- colnames(tmp)[hcl$order]
ha <- HeatmapAnnotation(
  bioproject = meta %>% column_to_rownames("sra") %>% .[colnames(tmp), "bioproject"],
  host = meta %>% column_to_rownames("sra") %>% .[colnames(tmp), "host"]
)

ht <- Heatmap(
  tmp_cor,
  name = "samples",
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  show_column_dend = FALSE,
  show_column_names = FALSE,
  show_row_names = FALSE,
  top_annotation = ha,
  height=nrow(tmp_cor) * unit(0.5, "mm"),
  width=ncol(tmp_cor) * unit(0.5, "mm")
)

rm(tmp_cor)
rm(tmp)

htsize <- calc_ht_size(ht)
htsize

pdf("output/06-samples_correlation_host.pdf", width = htsize[1], height = htsize[2])
ht
dev.off()

draw(ht)
```


There are a couple of good clusters there.
It looks like the Lettuce (Lactuca sativa) samples are all well correlated.
We'll see i guess.

## Differential gene expression


So i'm going to need to process each experiment separately.
Which means some quick PCA, dispersion, and applying the contrasts.
To save time i'll try to automate as much as I can.

We'll use DESeq2, which is my favourite as it handles genes with low counts better than EdgeR.

Here are the different bioprojects.

```{r}
unique(meta$bioproject)
```



```{r}
plotPCA2 <- function(dds, f1, f2) {
  pcaData <- plotPCA(
    dds,
    ntop = 1000,
    intgroup = c(f1, f2),
    returnData = TRUE
  )

  percentVar <- round(100 * attr(pcaData, "percentVar"))

  gg <- ggplot(pcaData, aes(x = PC1, y = PC2, color = .data[[f1]], shape = .data[[f2]])) +
    geom_point(size=3, alpha=0.75) +
    xlab(paste0("PC1: ", percentVar[1], "% variance")) +
    ylab(paste0("PC2: ", percentVar[2], "% variance"))
  
  gg <- gg_shape(gg, pcaData[[f2]])
  return(gg)
}

# Converts 0 to the smallest possible number before log calc
get_log <- function(vec){
  vec[vec == 0] <- .Machine$double.xmin
  vec <- -log10(vec)
}

reverselog_trans <- function(base = exp(1)) {
    trans <- function(x) -log(x, base)
    
    inv <- function(x) base^(-x)
    trans_new(paste0("reverselog-", format(base)), trans, inv, 
              log_breaks(base = base), 
              domain = c(1e-100, Inf))
}


plot_volcano <- function(df, title=NA, nlabels = 5, padj_threshold = 0.05) {
  df <- df[!is.na(df$padj), ]

  df_up <- df[(df$log2FoldChange > 0),]
  df_up <- df_up[order(df_up$padj),]

  df_up$label <- NA
  df_up$label[1:nlabels] <- rownames(df_up)[1:nlabels]

  df_down <- df[(df$log2FoldChange <= 0),]
  df_down <- df_down[order(df_down$padj),]
  df_down$label <- NA
  df_down$label[1:nlabels] <- rownames(df_down)[1:nlabels]

  df <- rbind(df_up, df_down)
  
  gg <- df %>%
    as.data.frame() %>%
    ggplot(aes(y=padj, x=log2FoldChange, colour=padj < padj_threshold, label = label)) +
    geom_text_repel(colour="black", box.padding=0.3, min.segment.length = 0, size=3) +
    geom_point(alpha=0.5) +
    scale_y_continuous(trans = reverselog_trans(10)) +
    ylab("adjusted p-value") +
    xlab(expression(paste(log[2] * " FC"))) +
    labs(colour=paste0("adj. p < ", padj_threshold)) +

  if ( !is.na(title) ) {
    gg <- gg + ggtitle(title)
  }  

  gg <- ggMarginal(gg, margins = "x", type="histogram", fill="lightgrey", xparams = list(bins = 100, linewidth = 0.25, fill = "black"), size=4)

  return(gg)
}


process_contrast_mat <- function(contrasts, this_meta) {
  bpr = unique(this_meta[["bioproject"]])
  stopifnot(all(this_meta[["bioproject"]] == bpr))

  colnames(contrasts) <- c("sample1", "sample2", "contrast_group")
  contrasts <- as.data.frame(contrasts)
  contrasts["contrast"] <- apply(contrasts, MARGIN = 1, FUN = function(x) {paste0(x["sample1"], "_vs_", x["sample2"])})
  contrasts["bioproject"] <- bpr
  contrasts <- dplyr::left_join(
    contrasts,
    this_meta %>% select(sample1 = sample, host1 = host_tidyname, infection_stage1 = infection_stage) %>% unique(),
    by="sample1"
  )
  
  contrasts <- dplyr::left_join(
    contrasts,
    this_meta %>% select(sample2 = sample, host2 = host_tidyname, infection_stage2 = infection_stage) %>% unique(),
    by="sample2"
  )
  
  contrasts <- contrasts[, c('contrast_group', 'bioproject', 'contrast', 'sample1', 'sample2', 'host1', 'host2', 'infection_stage1', 'infection_stage2')] %>%
    mutate(sample1 = as.factor(sample1), sample2 = as.factor(sample2), host1 = as.factor(host1), host2 = as.factor(host2), infection_stage1 = as.factor(infection_stage1), infection_stage2 = as.factor(infection_stage2))
  return(contrasts)
}


process_host_contrast_mat <- function(this_meta) {
  bpr = unique(this_meta[["bioproject"]])
  stopifnot(all(this_meta[["bioproject"]] == bpr))

  names <- unique(as.character(this_meta[["host_tidyname"]]))
  mat <- expand.grid(names, names) %>%
    filter(Var1 != "IV") %>%
    mutate(
      bioproject = bpr,
      contrast = paste0(bpr, "_", Var1, "vs", Var2),
      contrast_group = ifelse(Var2 == "IV", "IPvsIV", paste0(Var1, "vs", Var2)),
      infection_stage1 = as.factor(NA),
      infection_stage2 = as.factor(NA),
      resistance1 = as.factor(NA),
      resistance2 = as.factor(NA),
      virus1 = as.factor(NA),
      virus2 = as.factor(NA)
    ) %>%
    rename(
      Var1 = "host1",
      Var2 = "host2"
    ) %>%
    filter(host1 != host2) %>%
    select(contrast_group, bioproject, contrast, host1, host2, infection_stage1, infection_stage2, virus1, virus2, resistance1, resistance2)
  return(mat)
}


process_ipvsivtp_contrast_mat <- function(this_meta) {
  bpr = unique(this_meta[["bioproject"]])
  stopifnot(all(this_meta[["bioproject"]] == bpr))

  names <- unique(as.character(this_meta[["ipvsivtp"]]))

  mat <- expand.grid(names, names) %>%
    filter(Var1 != "IV", Var2 == "IV") %>%
    mutate(
      bioproject = bpr,
      contrast_group = paste0("IPvsIV_", gsub("^[^_]+_", "", Var1, perl = TRUE)),
      contrast = paste0(bpr, "_", Var1, "vs", Var2),
      infection_stage1 = as.factor(gsub("^[^_]+_", "", Var1, perl = TRUE)),
      infection_stage2 = as.factor(NA),
      resistance1 = as.factor(NA),
      resistance2 = as.factor(NA),
      virus1 = as.factor(NA),
      virus2 = as.factor(NA)
    ) %>%
    rename(
      Var1 = "host1",
      Var2 = "host2"
    ) %>%
    filter(host1 != host2) %>%
    select(contrast_group, bioproject, contrast, host1, host2, infection_stage1, infection_stage2, virus1, virus2, resistance1, resistance2)
  return(mat)
}


process_timepoint_contrast_mat <- function(this_meta) {
  bpr = unique(this_meta[["bioproject"]])
  stopifnot(all(this_meta[["bioproject"]] == bpr))

  names <- unique(as.character(this_meta[["host_tidyname"]]))
  names <- names[names != "IV"]
  infection_stage <- unique(as.character(this_meta[["infection_stage"]]))
  infection_stage <- infection_stage[!is.na(infection_stage)]
  
  mat <- expand.grid(names, infection_stage, infection_stage) %>%
    mutate(
      bioproject = bpr,
      contrast_group = paste0(Var2, "vs", Var3),
      contrast = paste0(bpr, "_", Var1, "_", Var2, "vs", Var3),
      host1 = Var1,
      host2 = Var1,
      resistance1 = as.factor(NA),
      resistance2 = as.factor(NA),
      virus1 = as.factor(NA),
      virus2 = as.factor(NA)
    ) %>%
    rename(
      Var2 = "infection_stage1",
      Var3 = "infection_stage2"
    ) %>%
    filter(
      contrast_group %in% c("MvsE", "LvsE", "LvsM", "MResvsERes", "LResvsERes", "LResvsMRes")
    ) %>%
    select(contrast_group, bioproject, contrast, host1, host2, infection_stage1, infection_stage2, virus1, virus2, resistance1, resistance2)
  return(mat)
}

process_resistance_contrast_mat <- function(this_meta) {
  bpr = unique(this_meta[["bioproject"]])
  stopifnot(all(this_meta[["bioproject"]] == bpr))

  names <- unique(as.character(this_meta[["host_tidyname"]]))
  names <- names[names != "IV"]
  resistance <- unique(as.character(this_meta[["resistance"]]))
  resistance <- resistance[!is.na(resistance)]
  
  mat <- expand.grid(names, resistance, resistance) %>%
    filter(
      Var2 == "Susc",
      Var3 == "Res"
    ) %>%
    mutate(
      bioproject = bpr,
      contrast_group = paste0(Var2, "vs", Var3),
      contrast = paste0(bpr, "_", Var1, "_", Var2, "vs", Var3),
      host1 = Var1,
      host2 = Var1,
      infection_stage1 = as.factor(NA),
      infection_stage2 = as.factor(NA),
      resistance1 = as.factor("Susc"),
      resistance2 = as.factor("Res"),
      virus1 = as.factor(NA),
      virus2 = as.factor(NA)
    ) %>%
    select(contrast_group, bioproject, contrast, host1, host2, infection_stage1, infection_stage2, virus1, virus2, resistance1, resistance2)
  return(mat)
}

process_virus_contrast_mat <- function(this_meta) {
  bpr = unique(this_meta[["bioproject"]])
  stopifnot(all(this_meta[["bioproject"]] == bpr))

  names <- unique(as.character(this_meta[["host_tidyname"]]))
  virus <- unique(as.character(this_meta[["virus"]]))
  virus <- virus[!is.na(virus)]

  mat <- expand.grid(names, virus, virus) %>%
    filter(
      Var2 != "VF",
      Var3 == "VF"
    ) %>%
    mutate(
      bioproject = bpr,
      contrast_group = paste0(Var2, "vs", Var3),
      contrast = paste0(bpr, "_", Var1, "_", Var2, "vs", Var3),
      host1 = Var1,
      host2 = Var1,
      infection_stage1 = as.factor(NA),
      infection_stage2 = as.factor(NA),
      resistance1 = as.factor(NA),
      resistance2 = as.factor(NA),
      virus1 = as.factor(Var2),
      virus2 = as.factor(Var3)
    ) %>%
    select(contrast_group, bioproject, contrast, host1, host2, infection_stage1, infection_stage2, virus1, virus2, resistance1, resistance2)
  return(mat)
}


run_deg_tests <- function(lcounts, meta, contrasts, design, kind) {
  bpr = unique(meta[["bioproject"]])
  stopifnot(all(meta[["bioproject"]] == bpr))
  stopifnot(kind %in% c("host", "infection_stage", "resistance", "virus", "pairwise", "ipvsivtp"))

  this_counts <- lcounts %>% tidyr::pivot_wider(id_cols = c(id), names_from = sra, values_from = count) %>%
    column_to_rownames("id") %>%
    as.matrix()

  this_meta <- meta %>% column_to_rownames("sra") %>% .[colnames(this_counts), ]
  
  stopifnot(all(rownames(this_meta) == colnames(this_counts)))
  
  dds <- DESeqDataSetFromMatrix(
    countData = this_counts,
    colData = this_meta,
    design = design
  )

  intgroup <- as.character(design[[2]])
  intgroup <- intgroup[!intgroup %in% c("+", "-", ":", "*")]
  print(plotPCA(rlog(dds, blind = FALSE), intgroup = intgroup))

  dds_results <- DESeq(dds, test = "Wald")
  filter <- HTSFilter(dds_results, plot = TRUE, s.len=100)$filteredData
  
  # Normally i prefer to run lfcThreshold=1, but the meta analysis methods require a uniform p-value
  # distribution, and thresholded tests yield a big peak around pvalue=1.
  
  if (kind == "host") {
    meta_col <- "host_tidyname"
    contrast_col1 <- "host1"
    contrast_col2 <- "host2"
  } else if (kind == "ipvsivtp") {
    meta_col <- "ipvsivtp"
    contrast_col1 <- "host1"
    contrast_col2 <- "host2"
  } else if (kind == "infection_stage") {
    meta_col <- "infection_stage"
    contrast_col1 <- "infection_stage1"
    contrast_col2 <- "infection_stage2"
  } else if (kind == "resistance") {
    meta_col <- "resistance"
    contrast_col1 <- "resistance1"
    contrast_col2 <- "resistance2"
  } else if (kind == "virus") {
    meta_col <- "virus"
    contrast_col1 <- "virus1"
    contrast_col2 <- "virus2"
  } else if (kind == "pairwise") {
    meta_col <- "sample"
    contrast_col1 <- "sample1"
    contrast_col2 <- "sample2"
  } else {
    stop("This shouldn't happen")
  }
  
  tests <- apply(
    contrasts,
    MARGIN = 1,
    FUN = function(x) {results(filter, contrast = c(meta_col, x[contrast_col1], x[contrast_col2]), alpha = 0.01, lfcThreshold = 0, independentFiltering=FALSE)}
  )
  names(tests) <- contrasts$contrast
  
  gg <- do.call(
    rbind,
    lapply(names(tests), FUN=function(n) {data.frame(contrast=n, pvalue=tests[[n]]$pvalue)})
  ) %>%
    ggplot(aes(x=pvalue)) +
    geom_histogram() +
    facet_wrap(vars(contrast))

  print(gg)

  mapper <- function(t) {
    t2 <- as.data.frame(t) %>% summarise(
      up = sum((log2FoldChange > 0) & (padj < 0.05), na.rm = TRUE),
      down = sum((log2FoldChange < 0) & (padj < 0.05), na.rm = TRUE),
      n = n()
    )
    return(t2)
  }
  
  print(do.call(rbind, lapply(tests, mapper)))
  
  tests <- do.call(
    rbind,
    lapply(
      names(tests),
      FUN = function(n) {as.data.frame(tests[[n]]) %>% rownames_to_column("id") %>% mutate(contrast = n)}
    )
  )

  tests <- right_join(contrasts, tests, by = "contrast") %>%
    rename(baseMean = "basemean", log2FoldChange = "lfc", lfcSE = "lfc_se")
  
  if (kind == "ipvsivtp") {
    tests <- tests %>%
      mutate(host1 = as.factor(gsub("_.+$", "", host1, perl = TRUE)))
  }
  return(tests)
}

```



## Pairwise tests

First off i'm going to run all of the pairwise tests of interest.
One of the main things we're after is invitro vs inplanta comparisons.
However, we don't have all timepoints in all experiments and the invitro samples do not usually have a time dimension, which makes marginal estimates difficult.
We'll look at the intersections between pairwise samples to decide what's good.

Honestly there are so many tests to run that it's hard to spend the time to describe the whole process in detail.
So i'm just going to burn through some code.


```{r}
pairwise_results <- list()
```

```{r}
bpr <- "PRJNA261444"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_12hpi", "PRJNA261444_SsclScl0205_periphery_IVPDA_stem_12hpi", "IPvsIV",
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_24hpi", "PRJNA261444_SsclScl0205_periphery_IVPDA_stem_24hpi", "IPvsIV",
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_48hpi", "PRJNA261444_SsclScl0205_periphery_IVPDA_stem_48hpi", "IPvsIV",
  "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_12hpi", "PRJNA261444_SsclScl0205_periphery_IVPDA_stem_12hpi", "IPvsIV",
  "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_24hpi", "PRJNA261444_SsclScl0205_periphery_IVPDA_stem_24hpi", "IPvsIV",
  "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_48hpi", "PRJNA261444_SsclScl0205_periphery_IVPDA_stem_48hpi", "IPvsIV",
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_24hpi", "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_12hpi", "MvsE",
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_48hpi", "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_12hpi", "LvsE",
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_48hpi", "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_24hpi", "LvsE",
  "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_24hpi", "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_12hpi", "MvsERes",
  "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_48hpi", "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_12hpi", "LvsERes",
  "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_48hpi", "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_24hpi", "LvsMRes",
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_12hpi", "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_12hpi", "SuscvsRes",
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_24hpi", "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_24hpi", "SuscvsRes",
  "PRJNA261444_SsclScl0205_lesion_PsatLifter_stem_48hpi", "PRJNA261444_SsclScl0205_lesion_PsatPI240515_stem_48hpi", "SuscvsRes"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```

```{r}
bpr <- "PRJNA327437"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_1hpi", "PRJNA327437_Sscl1980_colony_IVMMGlu", "IPvsIV",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_3hpi", "PRJNA327437_Sscl1980_colony_IVMMGlu", "IPvsIV",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_6hpi", "PRJNA327437_Sscl1980_colony_IVMMGlu", "IPvsIV",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_12hpi", "PRJNA327437_Sscl1980_colony_IVMMGlu", "IPvsIV",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_24hpi", "PRJNA327437_Sscl1980_colony_IVMMGlu", "IPvsIV",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_48hpi", "PRJNA327437_Sscl1980_colony_IVMMGlu", "IPvsIV",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_12hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_1hpi", "MvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_24hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_1hpi", "MvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_48hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_1hpi", "LvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_12hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_3hpi", "MvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_24hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_3hpi", "MvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_48hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_3hpi", "LvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_12hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_6hpi", "MvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_24hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_6hpi", "MvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_48hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_6hpi", "LvsE",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_48hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_12hpi", "LvsM",
  "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_48hpi", "PRJNA327437_Sscl1980_lesion_BnapDH12705_leaf_24hpi", "LvsM"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```

```{r}
bpr <- "PRJNA341340"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA341340_Sscl1980_lesion_Slys", "PRJNA341340_Sscl1980_ball_IVPDB", "IPvsIV",
  "PRJNA341340_Sscl1980_lesion_Hann", "PRJNA341340_Sscl1980_ball_IVPDB", "IPvsIV",
  "PRJNA341340_Sscl1980_lesion_Atha", "PRJNA341340_Sscl1980_ball_IVPDB", "IPvsIV",
  "PRJNA341340_Sscl1980_lesion_Slys", "PRJNA341340_Sscl1980_lesion_Hann", "SlysvsHann",
  "PRJNA341340_Sscl1980_lesion_Slys", "PRJNA341340_Sscl1980_lesion_Atha", "SlysvsAtha",
  "PRJNA341340_Sscl1980_lesion_Hann", "PRJNA341340_Sscl1980_lesion_Slys", "HannvsSlys",
  "PRJNA341340_Sscl1980_lesion_Hann", "PRJNA341340_Sscl1980_lesion_Atha", "HannvsAtha",  
  "PRJNA341340_Sscl1980_lesion_Atha", "PRJNA341340_Sscl1980_lesion_Slys", "AthavsSlys",
  "PRJNA341340_Sscl1980_lesion_Atha", "PRJNA341340_Sscl1980_lesion_Hann", "AthavsHann",  
  "PRJNA341340_Sscl1980_sclerotia_IV", "PRJNA341340_Sscl1980_ball_IVPDB", "Sclerotial"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA418121"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA418121_Sscl1980_periphery_AthaCol0_leaf_50hpi", "PRJNA418121_Sscl1980_periphery_SlysHeinz_leaf_50hpi", "AthavsSlys",
  "PRJNA418121_Sscl1980_periphery_SlysHeinz_leaf_50hpi", "PRJNA418121_Sscl1980_periphery_AthaCol0_leaf_50hpi", "SlysvsAtha"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA471709"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_48hpi", "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_24hpi", "MvsERes",
  "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_96hpi", "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_24hpi", "LvsERes",
  "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_96hpi", "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_48hpi", "LvsMRes",
  "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_48hpi", "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_24hpi", "MvsE",
  "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_96hpi", "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_24hpi", "LvsE",
  "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_96hpi", "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_48hpi", "LvsM",
  "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_24hpi", "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_24hpi", "SuscvsResE",
  "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_48hpi", "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_96hpi", "SuscvsRes",
  "PRJNA471709_Sscl1980_lesion_Gmax9144_stem_96hpi", "PRJNA471709_Sscl1980_lesion_Gmax91145_stem_96hpi", "SuscvsRes"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA477716"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA477716_Sscl1980_center_AthaCol0_leaf_48hpi", "PRJNA477716_Sscl1980_ball_IVPDB_96hpi", "IPvsIV",
  "PRJNA477716_Sscl1980_center_AthaCol0_leaf_48hpi", "PRJNA477716_Sscl1980_ball_IVMM_96hpi", "IPvsIV",
  "PRJNA477716_Sscl1980_center_AthaCol0_leaf_48hpi", "PRJNA477716_Sscl1980_periphery_IVPDA_48hpi", "IPvsIV",
  "PRJNA477716_Sscl1980_center_AthaCol0_leaf_48hpi", "PRJNA477716_Sscl1980_center_IVPDA_48hpi", "IPvsIV"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA501892"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_24hpi", "PRJNA501892_Sscl1980_colony_IVPDA", "IPvsIV",
  "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_48hpi", "PRJNA501892_Sscl1980_colony_IVPDA", "IPvsIV",
  "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_96hpi", "PRJNA501892_Sscl1980_colony_IVPDA", "IPvsIV",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_24hpi", "PRJNA501892_Sscl1980_colony_IVPDA", "IPvsIV",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_48hpi", "PRJNA501892_Sscl1980_colony_IVPDA", "IPvsIV",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_96hpi", "PRJNA501892_Sscl1980_colony_IVPDA", "IPvsIV",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_24hpi", "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_24hpi", "SuscvsResE",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_48hpi", "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_48hpi", "SuscvsRes",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_96hpi", "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_96hpi", "SuscvsRes",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_48hpi", "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_24hpi", "MvsE",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_96hpi", "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_24hpi", "LvsE",
  "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_96hpi", "PRJNA501892_Sscl1980_lesion_Gmax9144_stem_48hpi", "LvsM",
  "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_48hpi", "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_24hpi", "MvsERes",
  "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_96hpi", "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_24hpi", "LvsERes",
  "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_96hpi", "PRJNA501892_Sscl1980_lesion_Gmax91145_stem_48hpi", "LvsMRes"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```

```{r}
bpr <- "PRJNA516496"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA516496_SsclCU824_lesion_LangTanjil_stem_72hpi", "PRJNA516496_SsclCU824_ball_IVPDB", "IPvsIV",
  "PRJNA516496_SsclCU824_lesion_BnapCobbler_stem_72hpi", "PRJNA516496_SsclCU824_ball_IVPDB", "IPvsIV",
  "PRJNA516496_SsclCU824_lesion_LangTanjil_stem_72hpi", "PRJNA516496_SsclCU824_lesion_BnapCobbler_stem_72hpi", "LangvsBnap",
  "PRJNA516496_SsclCU824_lesion_BnapCobbler_stem_72hpi", "PRJNA516496_SsclCU824_lesion_LangTanjil_stem_72hpi", "BnapvsLang"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA574280"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA574280_Sscl1980_periphery_HannXRQ_leaf_24hpi", "PRJNA574280_Sscl1980_periphery_PvulG19833_leaf_50hpi", "HannvsPvul",
  "PRJNA574280_Sscl1980_periphery_Bvulvul_leaf_72hpi", "PRJNA574280_Sscl1980_periphery_PvulG19833_leaf_50hpi", "BvulvulvsPvul",
  "PRJNA574280_Sscl1980_periphery_Rcom_leaf_50hpi", "PRJNA574280_Sscl1980_periphery_PvulG19833_leaf_50hpi", "RcomvsPvul",
  "PRJNA574280_Sscl1980_periphery_PvulG19833_leaf_50hpi", "PRJNA574280_Sscl1980_periphery_HannXRQ_leaf_24hpi", "PvulvsHann",
  "PRJNA574280_Sscl1980_periphery_Bvulvul_leaf_72hpi", "PRJNA574280_Sscl1980_periphery_HannXRQ_leaf_24hpi", "BvulvulvsHann",
  "PRJNA574280_Sscl1980_periphery_Rcom_leaf_50hpi", "PRJNA574280_Sscl1980_periphery_HannXRQ_leaf_24hpi", "RcomvsHann",
  "PRJNA574280_Sscl1980_periphery_PvulG19833_leaf_50hpi", "PRJNA574280_Sscl1980_periphery_Bvulvul_leaf_72hpi", "PvulvsBvulvul",
  "PRJNA574280_Sscl1980_periphery_HannXRQ_leaf_24hpi", "PRJNA574280_Sscl1980_periphery_Bvulvul_leaf_72hpi", "HannvsBvulvul",
  "PRJNA574280_Sscl1980_periphery_Rcom_leaf_50hpi", "PRJNA574280_Sscl1980_periphery_Bvulvul_leaf_72hpi", "RcomvsBvulvul",
  "PRJNA574280_Sscl1980_periphery_PvulG19833_leaf_50hpi", "PRJNA574280_Sscl1980_periphery_Rcom_leaf_50hpi", "PvulvsRcom",
  "PRJNA574280_Sscl1980_periphery_HannXRQ_leaf_24hpi", "PRJNA574280_Sscl1980_periphery_Rcom_leaf_50hpi", "HannvsRcom",
  "PRJNA574280_Sscl1980_periphery_Bvulvul_leaf_72hpi", "PRJNA574280_Sscl1980_periphery_Rcom_leaf_50hpi", "BvulvulvsRcom"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA577619"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA577619_SsclDT8SsHADV_surface_hyphae_Bnap_12hpi", "PRJNA577619_SsclDT8_surface_hyphae_Bnap_12hpi", "SsHADVvsVF",
  "PRJNA577619_SsclDT8SsHADV_surface_hyphae_Bnap_24hpi", "PRJNA577619_SsclDT8_surface_hyphae_Bnap_24hpi", "SsHADVvsVF",
  "PRJNA577619_SsclDT8SsHADV_surface_hyphae_Bnap_18hpi", "PRJNA577619_SsclDT8_surface_hyphae_Bnap_18hpi", "SsHADVvsVF"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA593737"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA593737_SsclZ113OXSsNSRV1_colony_IVPDA", "PRJNA593737_Sscl1980_colony_IVPDA", "SsNRSVvsVF",
  "PRJNA593737_SsclZ11OXSsNSRV1_colony_IVPDA", "PRJNA593737_Sscl1980_colony_IVPDA", "SsNRSVvsVF",
  "PRJNA593737_SsclAH98_colony_IVPDA", "PRJNA593737_Sscl1980_colony_IVPDA", "SsNRSVSsHVvsVF"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA603456"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA603456_SsclMBGSs2_colony_IVPDAindol3carbinol", "PRJNA603456_SsclMBGSs2_colony_IVPDA", "none",
  "PRJNA603456_SsclMBGSs2_colony_IVPDAallylisothiocyanate", "PRJNA603456_SsclMBGSs2_colony_IVPDA", "none"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```

```{r}
bpr <- "PRJNA607858"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA607858_Sscl_lesion_AthaCol0Ago21_leaf_8hpi", "PRJNA607858_Sscl_lesion_AthaCol0_leaf_8hpi", "none"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA641217"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA641217_SsclDT8_lesion_Bnap_hypocotyl", "PRJNA641217_SsclDT8_colony_IVPDA", "IPvsIV",
  "PRJNA641217_SsclDT8SsHADV_lesion_Bnap_hypocotyl", "PRJNA641217_SsclDT8SsHADV_colony_IVPDA", "none",
  "PRJNA641217_SsclDT8SsHADV_colony_IVPDA", "PRJNA641217_SsclDT8_colony_IVPDA", "SsHADVvsVF",
  "PRJNA641217_SsclDT8SsHADV_lesion_Bnap_hypocotyl", "PRJNA641217_SsclDT8_lesion_Bnap_hypocotyl", "SsHADVvsVF"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA641462"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA641462_SsclDT8SsHADV_lesion_BnapHuashaung4_hypocotyl", "PRJNA641462_SsclDT8_lesion_BnapHuashaung4_hypocotyl", "SsHADVvsVF"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA643804"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  "PRJNA643804_SsclDK3SlaGemV1_cellophane_colony_IVPDACellophane", "PRJNA643804_SsclDK3_cellophane_colony_IVPDACellophane", "SlaGemV1vsVF"
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```

```{r}
bpr <- "PRJNA670487"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_ball_IVPDB', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_colony_IVPDA_DMSO', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_periphery_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_periphery_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_center_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_center_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_center_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_center_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_center_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_center_IVPDA', 'IPvsIV',
  'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'SlysvsAtha',
  'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'HannvsAtha',
  'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'BvulvulvsAtha',
  'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'RcomvsAtha',
  'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PvulvsAtha',
  'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'AthavsSlys',
  'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'HannvsSlys',
  'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'BvulvulvsSlys',
  'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'RcomvsSlys',
  'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PvulvsSlys',
  'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'AthavsHann',
  'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'SlysvsHann',
  'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'BvulvulvsHann',
  'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'RcomvsHann',
  'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PvulvsHann',
  'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'AthavsBvulvul',
  'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'SlysvsBvulvul',
  'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'HannvsBvulvul',
  'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'RcomvsBvulvul',
  'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PvulvsBvulvul',
  'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'AthavsRcom',
  'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'SlysvsRcom',
  'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'HannvsRcom',
  'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'BvulvulvsRcom',
  'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PvulvsRcom',
  'PRJNA670487_Sscl1980_periphery_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'AthavsPvul',
  'PRJNA670487_Sscl1980_periphery_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'SlysvsPvul',
  'PRJNA670487_Sscl1980_periphery_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'HannvsPvul',
  'PRJNA670487_Sscl1980_periphery_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'BvulvulvsPvul',
  'PRJNA670487_Sscl1980_periphery_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_periphery_Pvul_leaf_50hpi', 'RcomvsPvul',
  'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'SlysvsAtha',
  'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'HannvsAtha',
  'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'BvulvulvsAtha',
  'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'RcomvsAtha',
  'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PvulvsAtha',
  'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'AthavsSlys',
  'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'HannvsSlys',
  'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'BvulvulvsSlys',
  'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'RcomvsSlys',
  'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PvulvsSlys',
  'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'AthavsHann',
  'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'SlysvsHann',
  'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'BvulvulvsHann',
  'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'RcomvsHann',
  'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PvulvsHann',
  'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'AthavsBvulvul',
  'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'SlysvsBvulvul',
  'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'HannvsBvulvul',
  'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'RcomvsBvulvul',
  'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PvulvsBvulvul',
  'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'AthavsRcom',
  'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'SlysvsRcom',
  'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'HannvsRcom',
  'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'BvulvulvsRcom',
  'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PvulvsRcom',
  'PRJNA670487_Sscl1980_center_AthaCol0_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'AthavsPvul',
  'PRJNA670487_Sscl1980_center_Slys_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'SlysvsPvul',
  'PRJNA670487_Sscl1980_center_Hann_leaf_24hpi', 'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'HannvsPvul',
  'PRJNA670487_Sscl1980_center_Bvulvul_leaf_72hpi', 'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'BvulvulvsPvul',
  'PRJNA670487_Sscl1980_center_Rcom_leaf_50hpi', 'PRJNA670487_Sscl1980_center_Pvul_leaf_50hpi', 'RcomvsPvul'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA687280"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_6hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_12hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_24hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_48hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_72hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_72hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_48hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_24hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_12hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_6hpi', 'PRJNA687280_SsclCU820_ball_IVPDB', 'IPvsIV',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_6hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_6hpi', 'SuscvsResE',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_12hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_12hpi', 'SuscvsResE',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_24hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_24hpi', 'SuscvsRes',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_48hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_48hpi', 'SuscvsRes',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_24hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_6hpi', 'MvsERes',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_48hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_6hpi', 'MvsERes',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_72hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_6hpi', 'LvsERes',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_24hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_12hpi', 'MvsERes',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_48hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_12hpi', 'MvsERes',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_72hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_12hpi', 'LvsERes',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_72hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_24hpi', 'LvsMRes',
  'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_72hpi', 'PRJNA687280_SsclCU820_lesion_CariHatTrick_stem_48hpi', 'LvsMRes',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_24hpi', 'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_6hpi', 'MvsE',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_48hpi', 'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_6hpi', 'MvsE',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_72hpi', 'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_6hpi', 'LvsE',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_24hpi', 'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_12hpi', 'MvsE',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_48hpi', 'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_12hpi', 'MvsE',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_72hpi', 'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_12hpi', 'LvsE',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_72hpi', 'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_24hpi', 'LvsM',
  'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_72hpi', 'PRJNA687280_SsclCU820_lesion_CariKyabra_stem_48hpi', 'LvsM'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```

```{r}
bpr <- "PRJNA695466"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  'PRJNA695466_SsclDT8SsHADV_colony_IVPDA', 'PRJNA695466_SsclDT8_colony_IVPDA', 'SsHADVvsVF'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA706136"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  'PRJNA706136_Sscl_lesion_Bole_petiole_8hpi', 'PRJNA706136_Sscl_lesion_Bvil_petiole_8hpi', 'BolevsBvil',
  'PRJNA706136_Sscl_lesion_Bvil_petiole_8hpi', 'PRJNA706136_Sscl_lesion_Bole_petiole_8hpi', 'BvilvsBole'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA735329"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  'PRJNA735329_Sscl1980_lesion_BnapNingRS1_leaf_24hpi', 'PRJNA735329_Sscl1980_lesion_BnapNingRS1_leaf_6hpi', 'MvsERes',
  'PRJNA735329_Sscl1980_lesion_BnapNingRS1_leaf_48hpi', 'PRJNA735329_Sscl1980_lesion_BnapNingRS1_leaf_6hpi', 'LvsERes',
  'PRJNA735329_Sscl1980_lesion_BnapNingRS1_leaf_48hpi', 'PRJNA735329_Sscl1980_lesion_BnapNingRS1_leaf_24hpi', 'LvsMRes'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA766564"
this_meta <- meta %>% filter(bioproject == bpr, sra != "SRR16082437")
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
   'PRJNA766564_Sscl_lesion_BnapWestar_stem_168hpi', 'PRJNA766564_Sscl_lesion_BnapWestar_stem_24hpi', 'LvsE',
   'PRJNA766564_Sscl_lesion_BnapWestar_stem_168hpi', 'PRJNA766564_Sscl_lesion_Bnap1703_stem_168hpi', 'SuscvsRes'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA777355"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_2hpi', 'PRJNA777355_Sscl1980_cellophane_colony_IVPDACellophane', 'IPvsIV',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_6hpi', 'PRJNA777355_Sscl1980_cellophane_colony_IVPDACellophane', 'IPvsIV',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_12hpi', 'PRJNA777355_Sscl1980_cellophane_colony_IVPDACellophane', 'IPvsIV',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_24hpi', 'PRJNA777355_Sscl1980_cellophane_colony_IVPDACellophane', 'IPvsIV',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_36hpi', 'PRJNA777355_Sscl1980_cellophane_colony_IVPDACellophane', 'IPvsIV',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_12hpi', 'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_2hpi', 'MvsE',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_24hpi', 'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_2hpi', 'MvsE',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_36hpi', 'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_2hpi', 'MvsE',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_12hpi', 'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_6hpi', 'MvsE',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_24hpi', 'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_6hpi', 'MvsE',
  'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_36hpi', 'PRJNA777355_Sscl1980_lesion_BnapWestar_leaf_6hpi', 'MvsE'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```

```{r}
bpr <- "PRJNA789389"
this_meta <- meta %>% filter(bioproject == bpr, sra != "SRR17267260")
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  'PRJNA789389_SsclNB5_lesion_GmaxACColibriOA_leaf_4hpi', 'PRJNA789389_SsclNB5_ball_IVCYM', 'IPvsIV',
  'PRJNA789389_SsclNB5_lesion_GmaxACColibriOA_leaf_8hpi', 'PRJNA789389_SsclNB5_ball_IVCYM', 'IPvsIV',
  'PRJNA789389_SsclNB5_lesion_GmaxACColibri_leaf_8hpi', 'PRJNA789389_SsclNB5_ball_IVCYM', 'IPvsIV',
  'PRJNA789389_SsclNB5_lesion_GmaxACColibri_leaf_8hpi', 'PRJNA789389_SsclNB5_lesion_GmaxACColibriOA_leaf_8hpi', 'SuscvsResE'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r, fig.height=10, fig.width=10}
bpr <- "PRJNA804213"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

# These don't have phenotypes that I can find, so no contrasts
# PRJNA804213_SsclP7_lesion_LsatPI251246_leaf_42hpi
# PRJNA804213_SsclP7_lesion_LsatArmenianLser_leaf_42hpi
# PRJNA804213_SsclP7_lesion_Lsat43_leaf_42hpi

contrasts <- matrix(c(
  'PRJNA804213_SsclP7_lesion_LsatRedGranoble_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_Lsat58_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatPallone_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatOakLeaf_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatKaiserSelbstschluss_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatKahu_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatCobhamGreen_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatBloodyWarrior_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatAmbassador_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatAdriatica2_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_Lsat68_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatSimpson_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatSaladin_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatRomainDeBenicardo_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatPrazan_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatRedGranoble_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_Lsat58_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatPallone_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatOakLeaf_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatKaiserSelbstschluss_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatKahu_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatCobhamGreen_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatBloodyWarrior_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatAmbassador_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatAdriatica2_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_Lsat68_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatSimpson_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatSaladin_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatRomainDeBenicardo_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatIceberg_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatRedGranoble_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_Lsat58_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatPallone_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatOakLeaf_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatKaiserSelbstschluss_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatKahu_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatCobhamGreen_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatBloodyWarrior_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatAmbassador_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatAdriatica2_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_Lsat68_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatSimpson_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatSaladin_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatRomainDeBenicardo_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatBataviaTezier_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatRedGranoble_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_Lsat58_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatPallone_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatOakLeaf_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatKaiserSelbstschluss_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatKahu_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatCobhamGreen_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatBloodyWarrior_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatAmbassador_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatAdriatica2_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_Lsat68_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatSimpson_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatSaladin_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes',
  'PRJNA804213_SsclP7_lesion_LsatRomainDeBenicardo_leaf_42hpi', 'PRJNA804213_SsclP7_lesion_LsatAspen_leaf_42hpi', 'SuscvsRes'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```

```{r}
bpr <- "PRJNA830457"
this_meta <- meta %>% filter(bioproject == bpr, host_accession != "resynthesized")
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

# I decided to exclude the resynthesized sample in case it isn't a good representative of Bnapus
# PRJNA830457_Sscl_periphery_BnapResynthesized_leaf_24hpi

contrasts <- matrix(c(
  'PRJNA830457_Sscl_periphery_BrapIMB218_leaf_24hpi', 'PRJNA830457_Sscl_periphery_BoleTO1000_leaf_24hpi', 'BrapvsBole',
  'PRJNA830457_Sscl_periphery_BnapSentry_leaf_24hpi', 'PRJNA830457_Sscl_periphery_BoleTO1000_leaf_24hpi', 'BnapvsBole',
  'PRJNA830457_Sscl_periphery_BrapIMB218_leaf_24hpi', 'PRJNA830457_Sscl_periphery_BnapSentry_leaf_24hpi', 'BrapvsBnap',
  'PRJNA830457_Sscl_periphery_BoleTO1000_leaf_24hpi', 'PRJNA830457_Sscl_periphery_BnapSentry_leaf_24hpi', 'BolevsBnap',
  'PRJNA830457_Sscl_periphery_BnapSentry_leaf_24hpi', 'PRJNA830457_Sscl_periphery_BrapIMB218_leaf_24hpi', 'BnapvsBrap',
  'PRJNA830457_Sscl_periphery_BoleTO1000_leaf_24hpi', 'PRJNA830457_Sscl_periphery_BrapIMB218_leaf_24hpi', 'BolevsBrap'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
bpr <- "PRJNA874732"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])

contrasts <- matrix(c(
  'PRJNA874732_Sscl1980dSsAOX_colony_IVMMGluSSE', 'PRJNA874732_Sscl1980_colony_IVMMGluSSE', 'none'
), ncol=3, byrow = TRUE)

contrasts <- process_contrast_mat(contrasts, this_meta)
pairwise_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ sample, kind = "pairwise")
```


```{r}
pairwise_results_df <- do.call(rbind, pairwise_results)
rownames(pairwise_results_df) <- NULL

readr::write_tsv(pairwise_results_df, "output/dge_tests_pairwise.tsv", na = "-")
head(pairwise_results_df)
```

```{r, fig.width = 20, fig.height = 7}
tmp <- pairwise_results_df %>%
  filter(contrast_group == "IPvsIV", host1 == "Psat", padj < 0.01) %>%
  select(contrast, id) %>%
  split(., .[["contrast"]]) %>%
  lapply(FUN = function(df) {df[["id"]]})

upset(
  fromList(tmp),
  order.by = "freq",
  nsets = length(tmp),
  nintersects = 100
)
rm(tmp)
```

```{r, fig.width = 20, fig.height = 7}
tmp <- pairwise_results_df %>%
  filter(contrast_group %in% c("MvsE", "LvsE", "LvsM", "MvsERes", "LvsERes", "LvsMres"), host1 == "Psat", padj < 0.01) %>%
  select(contrast, id) %>%
  split(., .[["contrast"]]) %>%
  lapply(FUN = function(df) {df[["id"]]})

upset(
  fromList(tmp),
  order.by = "freq",
  nsets = length(tmp),
  nintersects = 100
)
rm(tmp)
```


```{r, fig.width = 20, fig.height = 7}
tmp <- pairwise_results_df %>%
  filter(contrast_group %in% c("MvsE", "LvsE", "LvsM", "MvsERes", "LvsERes", "LvsMres"), host1 == "Psat") %>%
  mutate(dge = as.integer(padj < 0.01)) %>%
  tidyr::pivot_wider(id_cols = "id", names_from = "contrast", values_from = "dge") %>%
  column_to_rownames("id")

tmp[is.na(tmp)] <- 0

heatmap(t(as.matrix(tmp)), labCol = FALSE, scale = "none", distfun = function(x) {dist(x, method = "binary")})
rm(tmp)
```

## Processing marginal sets

For the meta analysis, we really only want one DGE test for each experiment and contrast type of interest.
E.g. we would like to avoid having five ipvsiv contrasts, because the likely won't be independent of each other.
Here i'll try to select the best sets that we can get.


```{r}
host_results <- list()
ipvsiv_results <- list()
resistance_results <- list()
stage_results <- list()
virus_results <- list()
```

### PRJNA261444

Generally i'll test for susceptible vs resistance host differences.
If the have significantly different responses, i'll keep them separate.

```{r}
bpr <- "PRJNA261444"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(resistance))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_resistance_contrast_mat(this_meta)

resistance_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ resistance + hpi, kind = "resistance")
```

Because the resistant and susceptible cultivars are so similar, i won't distinguish between them for the marginal analysis of IPvsIV and infection stage.
In this case, because we have multiple time-points for the IV samples, i can do a two factor design.

```{r}
bpr <- "PRJNA261444"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname + hpi, kind = "host")
```

```{r}
bpr <- "PRJNA261444"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(infection_stage)) %>% mutate(infection_stage = factor(gsub("Res$", "", infection_stage)))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_timepoint_contrast_mat(this_meta)

stage_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ infection_stage + resistance, kind = "infection_stage")
```

```{r}
bpr <- "PRJNA261444"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(ipvsivtp)) %>% mutate(ipvsivtp = factor(gsub("Res$", "", ipvsivtp)))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_ipvsivtp_contrast_mat(this_meta)

ipvsiv_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ ipvsivtp, kind = "ipvsivtp")
```



### PRJNA327437

```{r}
bpr <- "PRJNA327437"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```

Here the factors of interest are IP vs IV, and 6 time points.


Because we don't have multiple timepoints for the invitro samples, we can't actually do a full marginal estimate.
The problem is that if we ignore the hpi, the variance will look very high and we'll only pick up things that are DE in all samples.
I think the best option will probably be to pick a "representative" time-point that's comparable across experiments?

```{r}
tmp <- lcounts %>% filter(host_tidyname == "Bnap") %>% tidyr::pivot_wider(id_cols = c(id), names_from = sra, values_from = cpm) %>%
  column_to_rownames("id") %>%
  as.matrix()

pca_raw <- princomp(tmp, fix_sign = TRUE)
rm(tmp)

prop_var <- ((pca_raw$sdev[1:10] / sum(pca_raw$sdev[1:10])) * 100) %>%
  data.frame(variance_explained = .) %>%
  rownames_to_column("PC") %>%
  mutate(PC = paste0("PC", sprintf("%02d", 1:n())))

print(ggplot(prop_var, aes(x=PC, y=variance_explained)) + geom_bar(stat = "identity"))

pca <- as.data.frame(pca_raw$loadings[, 1:3])
colnames(pca) <- paste0("PC", sprintf("%02d", seq_len(ncol(pca))))
pca <- pca %>% rownames_to_column("sra")
pca <- right_join(meta %>% select(bioproject, sra, hpi), pca, by = "sra") %>%
  tidyr::pivot_longer(-c(bioproject, sra, hpi), names_to = "PC")

gg <- ggplot(pca, aes(x=hpi, y=value, color=bioproject)) +
  geom_point(alpha = 0.7, size = 3) +
  facet_grid(cols = vars(PC))

print(gg)
```

Actually, it's pretty unclear what's "representative".
I think i'll just need to stick with processing inplanta vs invitro at each of our defined stages.
I think it's reasonable to pool some close time-points, and then at the end we can decide which of the three infection stages is most useful (probably mid or late though). 

```{r}
bpr <- "PRJNA327437"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```

```{r}
bpr <- "PRJNA327437"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(ipvsivtp))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_ipvsivtp_contrast_mat(this_meta)

ipvsiv_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ ipvsivtp, kind = "ipvsivtp")
```

```{r}
bpr <- "PRJNA327437"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(infection_stage)) %>% mutate(infection_stage = factor(gsub("Res$", "", infection_stage)))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_timepoint_contrast_mat(this_meta)

stage_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ infection_stage, kind = "infection_stage")
```


### PRJNA341340

```{r}
bpr <- "PRJNA341340"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA341340"
this_meta <- meta %>% filter(bioproject == bpr, tissue != "sclerotia")
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```


### PRJNA418121

```{r}
bpr <- "PRJNA418121"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```

```{r}
bpr <- "PRJNA418121"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```



### PRJNA471709


```{r}
bpr <- "PRJNA471709"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```

```{r}
bpr <- "PRJNA471709"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(resistance))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_resistance_contrast_mat(this_meta)

resistance_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ resistance + hpi, kind = "resistance")
```

There are enough differences in here between resistant and susceptible that I want to keep them separate.



```{r}
bpr <- "PRJNA471709"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(infection_stage))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_timepoint_contrast_mat(this_meta)

stage_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ infection_stage, kind = "infection_stage")
```


### PRJNA477716

```{r}
bpr <- "PRJNA477716"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, host_accession_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```

```{r}
bpr <- "PRJNA477716"
this_meta <- meta %>% filter(bioproject == bpr, host_accession_tidyname != "PDB", host_accession_tidyname != "MM")
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```



### PRJNA501892

```{r}
bpr <- "PRJNA501892"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```

```{r}
bpr <- "PRJNA501892"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(resistance))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_resistance_contrast_mat(this_meta)

resistance_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ resistance + hpi, kind = "resistance")
```

I think probably I should keep them separate.


```{r}
bpr <- "PRJNA501892"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(ipvsivtp))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_ipvsivtp_contrast_mat(this_meta)

ipvsiv_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ ipvsivtp, kind = "ipvsivtp")
```

```{r}
bpr <- "PRJNA501892"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```

```{r}
bpr <- "PRJNA501892"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(infection_stage))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_timepoint_contrast_mat(this_meta)

stage_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ infection_stage, kind = "infection_stage")
```



### PRJNA574280


```{r}
bpr <- "PRJNA574280"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA574280"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```




### PRJNA577619


```{r}
bpr <- "PRJNA577619"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA577619"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(virus))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_virus_contrast_mat(this_meta)

virus_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ virus + hpi, kind = "virus")
```


### PRJNA593737


```{r}
bpr <- "PRJNA593737"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA593737"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(virus), !sample %in% c("PRJNA593737_SsclZ11OXSsNSRV1_colony_IVPDA", "PRJNA593737_SsclZ113OXSsNSRV1_colony_IVPDA"))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_virus_contrast_mat(this_meta)

virus_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ virus, kind = "virus")
```



### PRJNA641217



```{r}
bpr <- "PRJNA641217"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA641217"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(virus))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_virus_contrast_mat(this_meta)

virus_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ virus + host_tidyname, kind = "virus")
```



### PRJNA641462


```{r}
bpr <- "PRJNA641462"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```

```{r}
bpr <- "PRJNA641462"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(virus))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_virus_contrast_mat(this_meta)

virus_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ virus, kind = "virus")
```



### PRJNA643804


```{r}
bpr <- "PRJNA643804"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA643804"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(virus))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_virus_contrast_mat(this_meta)

virus_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ virus, kind = "virus")
```





### PRJNA670487




```{r}
bpr <- "PRJNA670487"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA670487"
this_meta <- meta %>% filter(bioproject == bpr, sample != "PRJNA670487_Sscl1980_ball_IVPDB")
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```


### PRJNA687280



```{r}
bpr <- "PRJNA687280"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA687280"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(resistance))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_resistance_contrast_mat(this_meta)

resistance_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ resistance + hpi, kind = "resistance")
```

The PCA plot shows that the time-points have a much larger impact than the resistance or susceptibily.
But there are a few hundred DGEs, so i'll keep them separate.

```{r}
bpr <- "PRJNA687280"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```

```{r}
bpr <- "PRJNA687280"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(infection_stage))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_timepoint_contrast_mat(this_meta)

stage_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ infection_stage, kind = "infection_stage")
```

```{r}
bpr <- "PRJNA687280"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(ipvsivtp))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_ipvsivtp_contrast_mat(this_meta)

ipvsiv_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ ipvsivtp, kind = "ipvsivtp")
```









### PRJNA695466




```{r}
bpr <- "PRJNA695466"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA695466"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(virus))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_virus_contrast_mat(this_meta)

virus_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ virus, kind = "virus")
```



### PRJNA706136



```{r}
bpr <- "PRJNA706136"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```



```{r}
bpr <- "PRJNA706136"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```






### PRJNA735329


```{r}
bpr <- "PRJNA735329"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```

```{r}
# CHECK

bpr <- "PRJNA735329"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(infection_stage))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_timepoint_contrast_mat(this_meta)

stage_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ infection_stage, kind = "infection_stage")
```


### PRJNA766564




```{r}
bpr <- "PRJNA766564"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA766564"
this_meta <- meta %>% filter(bioproject == bpr, sample != "PRJNA766564_Sscl_lesion_Bnap1703_stem_24hpi") %>% mutate(resistance = ifelse(is.na(resistance), "Susc", "Res"))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_resistance_contrast_mat(this_meta)

resistance_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ resistance + hpi, kind = "resistance")
```

There's so little difference between resistance and susceptible that we won't worry about it.

```{r}
bpr <- "PRJNA766564"
this_meta <- meta %>%
  filter(bioproject == bpr, !is.na(infection_stage), sample != "PRJNA766564_Sscl_lesion_Bnap1703_stem_24hpi") %>%
  mutate(infection_stage = factor(gsub("Res$", "", infection_stage))) %>%
  mutate(resistance = ifelse(is.na(resistance), "Susc", "Res"))

this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_timepoint_contrast_mat(this_meta)

stage_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ infection_stage + resistance, kind = "infection_stage")
```



### PRJNA777355


```{r}
bpr <- "PRJNA777355"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA777355"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```

```{r}
bpr <- "PRJNA777355"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(infection_stage))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_timepoint_contrast_mat(this_meta)

stage_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ infection_stage, kind = "infection_stage")
```


```{r}
bpr <- "PRJNA777355"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(ipvsivtp))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_ipvsivtp_contrast_mat(this_meta)

ipvsiv_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ ipvsivtp, kind = "ipvsivtp")
```




### PRJNA789389


```{r}
bpr <- "PRJNA789389"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```



```{r}
bpr <- "PRJNA789389"
# Had to exclude sample, only one rep
this_meta <- meta %>% filter(bioproject == bpr, !is.na(resistance), sample != "PRJNA789389_SsclNB5_lesion_GmaxACColibri_leaf_4hpi")
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_resistance_contrast_mat(this_meta)

resistance_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ resistance + hpi, kind = "resistance")
```

```{r}
bpr <- "PRJNA789389"
this_meta <- meta %>% filter(bioproject == bpr)
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```


```{r}
bpr <- "PRJNA789389"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(ipvsivtp)) %>% mutate(ipvsivtp = factor(gsub("Res$", "", ipvsivtp)))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_ipvsivtp_contrast_mat(this_meta)

ipvsiv_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ ipvsivtp, kind = "ipvsivtp")
```




### PRJNA804213

```{r}
bpr <- "PRJNA804213"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA804213"
this_meta <- meta %>% filter(bioproject == bpr, !is.na(resistance))
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_resistance_contrast_mat(this_meta)

resistance_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ resistance, kind = "resistance")
```



### PRJNA830457

```{r}
bpr <- "PRJNA830457"
meta %>% filter(bioproject == bpr) %>% select(bioproject, sra, sample, host_tidyname, hpi, infection_stage, resistance, virus, ipvsivtp)
```


```{r}
bpr <- "PRJNA830457"
this_meta <- meta %>% filter(bioproject == bpr, sample != "PRJNA830457_Sscl_periphery_BnapResynthesized_leaf_24hpi")
this_lcounts <- lcounts %>% filter(sra %in% this_meta[["sra"]])
contrasts <- process_host_contrast_mat(this_meta)

host_results[[bpr]] <- run_deg_tests(this_lcounts, this_meta, contrasts, design = ~ host_tidyname, kind = "host")
```


Yay!
They're all done.
Now I need to combine the results and turn it into something we can use for the next steps.


```{r}
host_results_df <- do.call(rbind, host_results)
head(host_results_df)
```

```{r}
test <- host_results_df %>% mutate(dge = padj < 0.01) %>% filter(host2 == "IV") %>% tidyr::pivot_wider(id_cols = id, names_from = contrast, values_from = dge) %>%
  column_to_rownames("id") %>%
  as.matrix()

test[is.na(test)] <- FALSE
test <- test[!apply(test, MARGIN = 1, FUN = all), ]
test <- test[rowMeans(1.0 * test) > 0.1, ]
heatmap(1.0 * test, scale = "none")
```

```{r}
test <- host_results_df %>% mutate(dge = padj < 0.01) %>% filter(host2 != "IV") %>% tidyr::pivot_wider(id_cols = id, names_from = contrast, values_from = dge) %>%
  column_to_rownames("id") %>%
  as.matrix()

test[is.na(test)] <- FALSE
test <- test[!apply(test, MARGIN = 1, FUN = all), ]
test <- test[rowMeans(1.0 * test) > 0.1, ]
heatmap(1.0 * test, scale = "none")
```


```{r}
ipvsiv_results_df <- do.call(rbind, ipvsiv_results)
head(ipvsiv_results_df)
```

```{r}
test <- ipvsiv_results_df %>% mutate(dge = padj < 0.01) %>% tidyr::pivot_wider(id_cols = id, names_from = contrast, values_from = dge) %>%
  column_to_rownames("id") %>%
  as.matrix()

test[is.na(test)] <- FALSE
test <- test[!apply(test, MARGIN = 1, FUN = all), ]
test <- test[rowMeans(1.0 * test) > 0.1, ]
heatmap(1.0 * test, scale = "none")
```

```{r}
resistance_results_df <- do.call(rbind, resistance_results)
head(resistance_results_df)
```

```{r}
test <- resistance_results_df %>% mutate(dge = padj < 0.01) %>% tidyr::pivot_wider(id_cols = id, names_from = contrast, values_from = dge) %>%
  column_to_rownames("id") %>%
  as.matrix()

test[is.na(test)] <- FALSE
test <- test[!apply(test, MARGIN = 1, FUN = all), ]
test <- test[rowMeans(1.0 * test) > 0.1, ]
heatmap(1.0 * test, scale = "none")
```


```{r}
stage_results_df <- do.call(rbind, stage_results)
head(stage_results_df)
```

```{r}
test <- stage_results_df %>% mutate(dge = padj < 0.01) %>% tidyr::pivot_wider(id_cols = id, names_from = contrast, values_from = dge) %>%
  column_to_rownames("id") %>%
  as.matrix()

test[is.na(test)] <- FALSE
test <- test[!apply(test, MARGIN = 1, FUN = all), ]
test <- test[rowMeans(1.0 * test) > 0.1, ]
heatmap(1.0 * test, scale = "none")
```


```{r}
virus_results_df <- do.call(rbind, virus_results)
head(virus_results_df)
```


```{r}
test <- virus_results_df %>% mutate(dge = padj < 0.01) %>% tidyr::pivot_wider(id_cols = id, names_from = contrast, values_from = dge) %>%
  column_to_rownames("id") %>%
  as.matrix()

test[is.na(test)] <- FALSE
test <- test[!apply(test, MARGIN = 1, FUN = all), ]
test <- test[rowMeans(1.0 * test) > 0.1, ]
heatmap(1.0 * test, scale = "none")
```



```{r}
all_results <- list("ipvsiv" = ipvsiv_results_df, "infection_stage" = stage_results_df, "host" = host_results_df, "resistance" = resistance_results_df, "virus" = virus_results_df)
writexl::write_xlsx(all_results, "output/dge_tests.xlsx")
```


BYEEE!