---
title: "07-dge_metaanalysis.Rmd"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


In this notebook i'll be trying to combine the results of differential gene expression analyses to obtain more confident results.
Hopefully we'll be able to resolve some false positives and false negatives from individual studies.

To avoid re-doing work from the original papers, we'll focus on groups of DGE tests from multiple studies.

```{r}
library("tidyverse")
library("UpSetR")
library("metaRNASeq")
```

```{r}
ggplot2::theme_set(ggplot2::theme_bw() +
  ggplot2::theme(
    strip.background = element_rect(
     color="black", fill="white", linewidth=0, linetype="solid"
    ),
    rect = element_rect(
      colour = "black",
      linewidth = 1,
      linetype = "solid"
    )
  )
)

okabe <- c('#000000', '#E69F00', '#56B4E9', '#009E73', '#F0E042', '#0072B2', '#D55E00', '#CC79A7', '#999999')
options(ggplot2.discrete.fill = list(okabe[seq(2, 9, 2)], okabe[2:8], okabe, rep(okabe, 10)))
options(ggplot2.discrete.colour = list(okabe[seq(2, 9, 2)], okabe[2:8], okabe, rep(okabe, 10)))

gg_shape <- function(gg, vals) {gg + scale_shape_manual(values = rep(c(15, 17:20), 100))}
```


First i need to load the data.

```{r}
dge <- list()

for (sheetname in readxl::excel_sheets("output/dge_tests.xlsx") ) {
  dge[[sheetname]] <- readxl::read_xlsx("output/dge_tests.xlsx", sheet = sheetname)
}
head(dge[[1]])
```


## _in planta_ vs _in vitro_

We'll start with the most common kind of contrast, comparing infection vs media growth.

I'll look at these for each host species individually, and we'll probably have split those further as well.
So how many samples do we have for each species?


```{r}
dge[["host"]] %>% filter(contrast_group == "IPvsIV")%>% group_by(contrast_group, contrast, host1, host2) %>%
  summarise(up = sum((padj < 0.01) & (lfc > 0), na.rm = TRUE),  down = sum((padj < 0.01) & (lfc < 0), na.rm = TRUE)) %>%
  pivot_longer(cols = c("up", "down"), names_to = "direction", values_to = "count") %>%
  ggplot(aes(x = contrast, y = count, fill = direction)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(host1), drop = TRUE, scales = "free") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```


```{r}
dge[["ipvsiv"]] %>% group_by(contrast_group, contrast, host1, host2, infection_stage1) %>%
  summarise(up = sum((padj < 0.01) & (lfc > 0), na.rm = TRUE),  down = sum((padj < 0.01) & (lfc < 0), na.rm = TRUE)) %>%
  pivot_longer(cols = c("up", "down"), names_to = "direction", values_to = "count") %>%
  ggplot(aes(x = contrast, y = count, fill = direction)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(host1), drop = TRUE, scales = "free") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

Right, so with the IP vs IV samples we need to decide for each bioproject/host pair whether we need to use the group-wise option or a specific time-point.

These are all of the hosts we need to look at.

```{r}
ipvsiv <- dge[["host"]] %>% filter(host2 == "IV")
ipvsiv %>% .[["host1"]] %>% unique()
```


```{r}
ipvsiv2 <- dge[["ipvsiv"]]
ipvsiv2 %>% .[["host1"]] %>% unique()
```

```{r}
ipvsiv_curated <- list()

ipvsiv_curated[["rest"]] <- ipvsiv %>% filter(!host1 %in% c("Psat", "Bnap", "Gmax", "Cari"))
```


```{r}
ipvsiv %>% filter(host1 == "Psat") %>% .[["bioproject"]] %>% unique()
```

```{r}
ipvsiv2 %>% filter(host1 == "Psat") %>% select(bioproject, infection_stage1) %>% unique()
```

ok so for Pea we have only one bioproject with ip vs iv, and it had time-points of in vitro samples.
We'll take the marginal.

```{r}
ipvsiv_curated[["Psat"]] <- ipvsiv %>% filter(host1 == "Psat")
```



```{r}
ipvsiv %>% filter(host1 == "Bnap") %>% .[["bioproject"]] %>% unique()
```


```{r}
ipvsiv2 %>% filter(host1 == "Bnap") %>% select(bioproject, infection_stage1) %>% unique()
```

For brassica napus we have two time courses and i couln't get proper marginal contrasts.
We'll use the mid stage time points, 12H-24H

```{r}
ipvsiv_curated[["Bnap"]] <- ipvsiv2 %>% filter(host1 == "Bnap", infection_stage1 == "M")
```


```{r}
ipvsiv %>% filter(host1 == "Gmax") %>% .[["bioproject"]] %>% unique()
```

```{r}
ipvsiv2 %>% filter(host1 == "Gmax") %>% select(bioproject, infection_stage1) %>% unique()
```
For soybean because we only have the early time point in both sets we'll take that.

```{r}
ipvsiv_curated[["Gmax"]] <- ipvsiv2 %>% filter(host1 == "Gmax", infection_stage1 == "E")
```

```{r}
ipvsiv %>% filter(host1 == "Cari") %>% .[["bioproject"]] %>% unique()
```

```{r}
ipvsiv2 %>% filter(host1 == "Cari") %>% select(bioproject, infection_stage1) %>% unique()
```

```{r}
ipvsiv_curated[["Cari"]] <- ipvsiv2 %>% filter(host1 == "Cari", infection_stage1 == "L")
```


```{r}
ipvsiv <- do.call(rbind, ipvsiv_curated)
ipvsiv["contrast_group"] <- "IPvsIV"
ipvsiv
```




```{r, fig.width = 20, fig.height=10}
process_dge <- function(dge, set = NULL, padj_threshold = 0.01, sign_threshold = 0.9) {
  mask <- rep(TRUE, nrow(dge))
  
  for (n in names(set)) {
    mask <- mask & (dge[[n]] == set[[n]])
  }

  pmat <- dge[mask, ] %>%
    pivot_wider(id_cols = "id", names_from = "contrast", values_from = "pvalue") %>%
    column_to_rownames("id") %>%
    as.data.frame()
  
  clist <- lapply(colnames(pmat), FUN = function(n) {rownames(pmat)[(pmat[, n] < padj_threshold) & (!is.na(pmat[, n]))]})
  names(clist) <- colnames(pmat)
  
  gg <- dge[mask, ] %>%
    ggplot(aes(x = pvalue)) +
    geom_histogram(bins = 100) +
    facet_wrap(vars(contrast), drop = TRUE) +
    theme(axis.title.x=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks.x=element_blank())

  print(gg)

  
  stopifnot(FALSE)
  I want all significant tests to have the same sign, instead of what we're doing now
  lfc <- dge[mask, ] %>%
    mutate(sign_total = sign(lfc), sign_sig = ifelse(padj < padj_threshold, sign(lfc), NA)) %>%
    group_by(id) %>%
    summarise(
      lfc_mean = ifelse(all(is.na(lfc)), NA, mean(lfc, na.rm = TRUE)),
      lfc_median = ifelse(all(is.na(lfc)), NA, median(lfc, na.rm = TRUE)),
      lfc_iqr = ifelse(all(is.na(lfc)), NA, IQR(lfc, na.rm = TRUE)),
      lfc_min = ifelse(all(is.na(lfc)), NA, min(lfc, na.rm = TRUE)),
      lfc_max = ifelse(all(is.na(lfc)), NA, max(lfc, na.rm = TRUE)),
      sign_total = ifelse(all(is.na(lfc)), NA, sum(sign_total, na.rm = TRUE)),
      ntot = n(),
      sign_sig = sum(sign_sig, na.rm = TRUE),
      nsig = sum(abs(sign_sig), na.rm = TRUE),
      consistent_sign = ((abs(sign_total) / ntot >= sign_threshold) & (abs(sign_sig) == nsig))
    )
  
  fishcomb <- fishercomb(pmat, BHth = padj_threshold)
  
  hist(fishcomb$rawpval)
  
  adjpval <- fishcomb$adjpval
  names(adjpval) <- rownames(pmat)
  
  lfc["meta_padj"] <- adjpval[lfc[["id"]]]
  
  clist[["meta"]] <- lfc %>% filter(!is.na(meta_padj), (meta_padj < padj_threshold) & consistent_sign) %>% select(id) %>% .[[1]]
  p <- upset(
    fromList(clist),
    order.by = "freq",
    nsets = length(clist),
    nintersects = 30
  )

  print(p)
  
  gg <- ggplot(
    lfc %>% mutate(significant = consistent_sign & (meta_padj < padj_threshold)),
    aes(x=lfc_median, y=-log10(meta_padj))
  ) +
    geom_point(aes(x = ifelse(.data[["significant"]], NA, .data[["lfc_median"]]), y = ifelse(.data[["significant"]], NA, -log10(.data[["meta_padj"]]))), alpha = 0.2, size = 1.5, colour = "#E69F00") +
    geom_point(aes(x = ifelse(.data[["significant"]], .data[["lfc_median"]], NA), y = ifelse(.data[["significant"]], -log10(.data[["meta_padj"]]), NA)), alpha = 0.75, size = 1.5, colour = "black")

  print(gg)

  if (!is.null(set)) {
    for (n in names(set)) {
      lfc[n] <- set[[n]] 
    }
  }
    
  for (n in names(clist)) {
    lfc[n] <- lfc[["id"]] %in% clist[[n]]
  }

  
  return(lfc)
}
```


```{r}
all_dges <- list()

ipvsiv_contrasts <- ipvsiv %>% select(host1, host2) %>% unique()
for (i in 1:nrow(ipvsiv_contrasts)) {
  row <- unlist(ipvsiv_contrasts[i,])
  print("")
  print("#####################################################################################")
  print(row)

  dge_matches <- process_dge(ipvsiv, list("contrast_group" = "IPvsIV", "host1" = row["host1"], "host2" = row["host2"]))
  dge_matches[is.na(dge_matches$meta), "meta"] <- FALSE
  all_dges[[paste0("IPvsIV_", row["host1"])]] <- dge_matches
}

writexl::write_xlsx(all_dges, path = "output/ipvsiv_meta.xlsx")
```

Ok. So we've run all of them.
Let's look at the overall comparisons.

```{r, fig.height=6, fig.width=8}
ipvsiv_meta <- lapply(all_dges, FUN = function(df) {
  if ("bioproject" %in% colnames(df)) {
    r <- df %>% select(id, meta, host1, host2, bioproject)
  } else {
    r <- df %>% select(id, meta, host1, host2) %>% mutate(bioproject = NA)
  }
  return(r)
}) %>% do.call(rbind, .)

ipvsiv_meta["contrast"] <- ifelse(is.na(ipvsiv_meta$bioproject), paste0(ipvsiv_meta$host1, "_vs_", ipvsiv_meta$host2), paste0(ipvsiv_meta$host1, "_vs_", ipvsiv_meta$host2, "_", ipvsiv_meta$bioproject)) 

ipvsiv_meta <- ipvsiv_meta %>% pivot_wider(id_cols = "id", names_from = "contrast", values_from = "meta") %>% column_to_rownames("id")
ipvsiv_meta[is.na(ipvsiv_meta)] <- FALSE

clist <- lapply(colnames(ipvsiv_meta), FUN = function(col) {rownames(ipvsiv_meta)[ipvsiv_meta[[col]]]})
names(clist) <- colnames(ipvsiv_meta)

p <- upset(
  fromList(clist),
  order.by = "freq",
  nsets = length(clist),
  nintersects = 50
)
print(p)
```

```{r, fig.height=8, fig.width=8}
d <- dist(t(ipvsiv_meta[rowSums(ipvsiv_meta) > 1, ]), method = "binary")
hcl <- hclust(d)
plot(hcl)
#heatmap(apply(ipvsiv_meta[rowSums(ipvsiv_meta) > 0, ], MARGIN=2, as.numeric), scale = "none")
```


There are ~-80 genes that are DGE in all IPvsIV samples.
What's going on there?


```{r}
funcs <- readr::read_tsv("input/predicted_protein_functions.tsv")

DT::datatable(funcs %>% filter(name %in% rownames(ipvsiv_meta[apply(ipvsiv_meta, MARGIN = 1, FUN = all), ])) %>% select(name, effector_score, effectordb_name, phibase_genes, dbcan_description, swissprot_protein_name, pdb_protein_name, ipr_name))
```

Unsurpisingly, most things are CAZymes, but there are a few other things in there that are interesting.
Including 1 known sclero effector and 1 homologue of BghBEC2.

The homologue could be a lot of things, BEC2 is thought to interfere with host resistance.
It's also homologous to some Magnaporthe CDIPs (cell death inducing proteins).
And interestingly, it has similarity to a Copsin protein, which are antimicrobial.



## Time points.

```{r}
infection_stage <- dge$infection_stage
head(infection_stage)
```

Let's look at some time-points!
I'll start with the mid vs early time points

```{r}
infection_stage %>% filter(contrast_group == "MvsE") %>% group_by(contrast_group, contrast, host1) %>%
  summarise(up = sum((padj < 0.01) & (lfc > 0), na.rm = TRUE),  down = sum((padj < 0.01) & (lfc < 0), na.rm = TRUE)) %>%
  pivot_longer(cols = c("up", "down"), names_to = "direction", values_to = "count") %>%
  ggplot(aes(x = contrast, y = count, fill = direction)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(host1), drop = TRUE, scales = "free") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

```{r}
infection_stage %>% filter(contrast_group == "LvsE") %>% group_by(contrast_group, contrast, host1) %>%
  summarise(up = sum((padj < 0.01) & (lfc > 0), na.rm = TRUE),  down = sum((padj < 0.01) & (lfc < 0), na.rm = TRUE)) %>%
  pivot_longer(cols = c("up", "down"), names_to = "direction", values_to = "count") %>%
  ggplot(aes(x = contrast, y = count, fill = direction)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(host1), drop = TRUE, scales = "free") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

```{r}
infection_stage %>% filter(contrast_group == "LvsM") %>% group_by(contrast_group, contrast, host1) %>%
  summarise(up = sum((padj < 0.01) & (lfc > 0), na.rm = TRUE),  down = sum((padj < 0.01) & (lfc < 0), na.rm = TRUE)) %>%
  pivot_longer(cols = c("up", "down"), names_to = "direction", values_to = "count") %>%
  ggplot(aes(x = contrast, y = count, fill = direction)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(host1), drop = TRUE, scales = "free") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

Right, so there are 4 species with time-series information.
Most only have two different experiments (which is still enough to do something with).


```{r, fig.height = 10, fig.width = 15}
all_dges2 <- list()
#%>% filter(contrast_group %in% c("MvsE", "LvsE", "LvsM"))
timepoint_contrasts <- infection_stage  %>% select(contrast_group, host1) %>% unique()

for (i in 1:nrow(timepoint_contrasts)) {
  row <- unlist(timepoint_contrasts[i,])
  print("")
  print("#####################################################################################")
  print(row)
  
  #if ((row["host1"] == "Bnap") & any("PRJNA327437" %in% dge[(dge$contrast_group == row["contrast_group"]) & (dge$host1 == "Bnap"), ]$bioproject)) {
  #  dge_matches <- process_dge(dge, list("contrast_group" = row["contrast_group"], "host1" = row["host1"], "bioproject" = "PRJNA327437"))
  #  all_dges2[[paste(row["contrast_group"], row["host1"], "PRJNA327437", sep="_")]] <- dge_matches
  #
  #  tmp <- dge %>% filter(bioproject != "PRJNA327437", contrast_group == row["contrast_group"], host1 == row["host1"])
  #  if (nrow(tmp) > 0) {
  #    dge_matches <- process_dge(tmp, list("contrast_group" = row["contrast_group"], "host1" = row["host1"]))
  #    all_dges2[[paste(row["contrast_group"], row["host1"], sep="_")]] <- dge_matches
  #  }
  #} else {
  dge_matches <- process_dge(infection_stage, list("contrast_group" = row["contrast_group"], "host1" = row["host1"]))
  all_dges2[[paste(row["contrast_group"], row["host1"], sep="_")]] <- dge_matches
}

writexl::write_xlsx(all_dges2, path = "output/timepoint_meta.xlsx")
```

I don't know that we'll look super in depth at the time-points WRT resistant cultivars, but i've got them here for completeness.

Ok. So we've run all of them.
Let's look at the overall comparisons.

```{r, fig.height=9, fig.width=10}
timepoints <- lapply(all_dges2, FUN = function(df) {
  if ("bioproject" %in% colnames(df)) {
    r <- df %>% select(id, meta, contrast_group, host1, bioproject)
  } else {
    r <- df %>% select(id, meta, contrast_group, host1) %>% mutate(bioproject = NA)
  }
  return(r)
}) %>% do.call(rbind, .)

timepoints["contrast"] <- ifelse(is.na(timepoints$bioproject), paste0(timepoints$contrast_group, "_", timepoints$host1), paste0(timepoints$contrast_group, "_", timepoints$host1, "_", timepoints$bioproject))

timepoints <- timepoints %>% pivot_wider(id_cols = "id", names_from = "contrast", values_from = "meta") %>% column_to_rownames("id")

clist <- lapply(colnames(timepoints), FUN = function(col) {rownames(timepoints)[timepoints[[col]]]})
names(clist) <- colnames(timepoints)

p <- upset(
  fromList(clist),
  order.by = "freq",
  nsets = length(clist),
  nintersects = 50
)
print(p)
```


```{r, fig.height=8, fig.width=8}
d <- dist(t(ipvsiv_meta[rowSums(ipvsiv_meta) > 1, ]), method = "binary")
hcl <- hclust(d)
plot(hcl)
#heatmap(apply(ipvsiv_meta[rowSums(ipvsiv_meta) > 0, ], MARGIN=2, as.numeric), scale = "none")
```

```{r, fig.height=8, fig.width=8}
d <- dist(t(apply(timepoints[rowSums(timepoints) > 0, ], MARGIN = 2, FUN = as.numeric)), method = "binary")
hcl <- hclust(d)
plot(hcl)
#heatmap(apply(timepoints[rowSums(timepoints) > 0, ], MARGIN=2, as.numeric), scale = "none")
```

What i'm seeing is that generally there isn't much of a difference between Late vs early or mid infection.
Probably late infection just has a lot going on (and more transcripts, so less bias).

There's some consistency in how they group together, but generally within the same time-point and host.


the Bnapus resistant samples are quite interesting.
For most of the other timepoints comparing resistant samples they group with the susceptible cultivar.
But it seems like the Brassicas actually do have a different response, and it's not just similar to an earlier time-point.


## Compare susceptible hosts vs resistant ones.

```{r}
resistance <- dge$resistance
head(resistance)
```


```{r}
resistance %>% group_by(contrast_group, contrast, host1) %>%
  summarise(up = sum((padj < 0.01) & (lfc > 0), na.rm = TRUE),  down = sum((padj < 0.01) & (lfc < 0), na.rm = TRUE)) %>%
  pivot_longer(cols = c("up", "down"), names_to = "direction", values_to = "count") %>%
  ggplot(aes(x = contrast, y = count, fill = direction)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(host1), drop = TRUE, scales = "free") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

```{r}
all_dges3 <- list()
```


```{r}
dge_matches <- process_dge(resistance, list("contrast_group" = "SuscvsRes", "host1" = "Bnap"))
all_dges3[["Bnap"]] <- dge_matches

dge_matches <- process_dge(resistance, list("contrast_group" = "SuscvsRes", "host1" = "Gmax"))
all_dges3[["Gmax"]] <- dge_matches

dge_matches <- process_dge(resistance, list("contrast_group" = "SuscvsRes", "host1" = "Cari"))
all_dges3[["Cari"]] <- dge_matches

dge_matches <- process_dge(resistance, list("contrast_group" = "SuscvsRes", "host1" = "Lsat"))
all_dges3[["Lsat"]] <- dge_matches

# NO DGEs, so this fails if you run it.
#dge_matches <- process_dge(resistance, list("contrast_group" = "SuscvsRes", "host1" = "Psat"))
#all_dges3[["Psat"]] <- dge_matches
```


Ok. So we've run all of them.
Let's look at the overall comparisons.

```{r}
names(all_dges3) <- unname(unlist(lapply(all_dges3, FUN = function(df) {
  if ("bioproject" %in% colnames(df)) {
    n <- paste0(df[1, c("contrast_group", "host1", "bioproject")], collapse="_")
  } else {
    n <- paste0(df[1, c("contrast_group", "host1")], collapse="_")
  }
  return(n)
})))

writexl::write_xlsx(all_dges3, path = "output/susc_vs_res_meta.xlsx")
```


```{r}
suscvsres <- lapply(all_dges3, FUN = function(df) {
  if ("bioproject" %in% colnames(df)) {
    r <- df %>% select(id, meta, contrast_group, host1, bioproject)
  } else {
    r <- df %>% select(id, meta, contrast_group, host1) %>% mutate(bioproject = NA)
  }
  return(r)
}) %>% do.call(rbind, .)

suscvsres["contrast"] <- ifelse(is.na(suscvsres$bioproject), paste0(suscvsres$contrast_group, "_", suscvsres$host1), paste0(suscvsres$contrast_group, "_", suscvsres$host1, "_", suscvsres$bioproject))

suscvsres <- suscvsres %>% pivot_wider(id_cols = "id", names_from = "contrast", values_from = "meta") %>% column_to_rownames("id")

clist <- lapply(colnames(suscvsres), FUN = function(col) {rownames(suscvsres)[suscvsres[[col]]]})
names(clist) <- colnames(suscvsres)

p <- upset(
  fromList(clist),
  order.by = "freq",
  nsets = length(clist),
  nintersects = 50
)
print(p)
```

Weirdly, the brassica napus has lost it's interesting Susceptibility vs Resistance differences that we saw in in the time-points...


```{r}
#heatmap(apply(suscvsres[rowSums(suscvsres) > 0, ], MARGIN=2, as.numeric), scale = "none")
```



### Host vs Host


```{r}
hosts <- dge$host %>% filter(contrast_group  != "IPvsIV")

hosts %>% group_by(contrast_group, contrast, host1, host2) %>%
  summarise(up = sum((padj < 0.01) & (lfc > 0), na.rm = TRUE),  down = sum((padj < 0.01) & (lfc < 0), na.rm = TRUE)) %>%
  pivot_longer(cols = c("up", "down"), names_to = "direction", values_to = "count") %>%
  ggplot(aes(x = contrast, y = count, fill = direction)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(contrast_group), drop = TRUE, scales = "free") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```


```{r, fig.width = 15, fig.height=10}
all_dges4 <- list()
hosts
for (l in unique(hosts$contrast_group)) {
  print("")
  print("######################################################################################")
  print(l)
  n <- strsplit(l, "vs")[[1]]
  dge_matches <- process_dge(hosts, list("contrast_group" = l, "host1" = n[1], "host2" = n[2]))
  all_dges4[[l]] <- dge_matches
}
writexl::write_xlsx(all_dges4, path = "output/hostvshost_meta.xlsx")
```
Ok. So we've run all of them.
Let's look at the overall comparisons.

```{r, fig.height=15, fig.width=12}
hostvshost <- lapply(all_dges4, FUN = function(df) {
  if ("bioproject" %in% colnames(df)) {
    r <- df %>% select(id, meta, contrast_group, host1, host2, bioproject)
  } else {
    r <- df %>% select(id, meta, contrast_group, host1, host2) %>% mutate(bioproject = NA)
  }
  return(r)
}) %>% do.call(rbind, .)

hostvshost["contrast"] <- ifelse(is.na(hostvshost$bioproject), hostvshost$contrast_group, paste0(hostvshost$contrast_group, "_", hostvshost$bioproject))

hostvshost <- hostvshost %>% pivot_wider(id_cols = "id", names_from = "contrast", values_from = "meta") %>% column_to_rownames("id")

clist <- lapply(colnames(hostvshost), FUN = function(col) {rownames(hostvshost)[hostvshost[[col]]]})
names(clist) <- colnames(hostvshost)

p <- upset(
  fromList(clist),
  order.by = "freq",
  nsets = length(clist),
  nintersects = 50
)
print(p)
```

```{r, fig.height=8, fig.width=8}
d <- dist(t(apply(hostvshost[rowSums(timepoints) > 1, ], MARGIN = 2, FUN = as.numeric)), method = "binary")
hcl <- hclust(d)
plot(hcl)

#heatmap(apply(hostvshost[rowSums(hostvshost) > 0, ], MARGIN=2, as.numeric), scale = "none")
```




## Virus

```{r, fig.width = 15, fig.height=10}
virus <- dge$virus

all_dges5 <- list()

for (l in unique(virus$contrast_group)) {
  print("")
  print("######################################################################################")
  print(l)

  dge_matches <- process_dge(virus, list("contrast_group" = l))
  all_dges5[[l]] <- dge_matches
}

writexl::write_xlsx(all_dges5, path = "output/virus_meta.xlsx")
```

I find it a bit odd that the SsHADV invitro and inplanta samples have such similar DGE profiles.
I guess the point that they make is that the fungus seems to lose virulence on hosts, but they also claim that it can grow endophytically in some monocots.
I guess I'd expect to still see some consistent differences between IV and IP contrasts.


```{r, fig.height=8, fig.width=10}
virus <- lapply(all_dges5, FUN = function(df) {
  if ("bioproject" %in% colnames(df)) {
    r <- df %>% select(id, meta, contrast_group, bioproject)
  } else {
    r <- df %>% select(id, meta, contrast_group) %>% mutate(bioproject = NA)
  }
  return(r)
}) %>% do.call(rbind, .)

virus["contrast"] <- virus$contrast_group

virus <- virus %>% pivot_wider(id_cols = "id", names_from = "contrast", values_from = "meta") %>% column_to_rownames("id")

clist <- lapply(colnames(virus), FUN = function(col) {rownames(virus)[virus[[col]]]})
names(clist) <- colnames(virus)

p <- upset(
  fromList(clist),
  order.by = "freq",
  nsets = length(clist),
  nintersects = 50
)
print(p)
```


