---
title: "08-coexpression.Rmd"
output: html_document
date: "2023-07-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library("tidyverse")
library("UpSetR")
library("igraph")
library("ComplexHeatmap")
library("ica")
```

```{r}
ggplot2::theme_set(ggplot2::theme_bw() +
  ggplot2::theme(
    strip.background = element_rect(
     color="black", fill="white", linewidth=0, linetype="solid"
    ),
    rect = element_rect(
      colour = "black",
      linewidth = 1,
      linetype = "solid"
    )
  )
)

okabe <- c('#000000', '#E69F00', '#56B4E9', '#009E73', '#F0E042', '#0072B2', '#D55E00', '#CC79A7', '#999999')
options(ggplot2.discrete.fill = list(okabe[seq(2, 9, 2)], okabe[2:8], okabe, rep(okabe, 10)))
options(ggplot2.discrete.colour = list(okabe[seq(2, 9, 2)], okabe[2:8], okabe, rep(okabe, 10)))

gg_shape <- function(gg) {gg + scale_shape_manual(values = rep(c(15, 17:20), 100))}
```

```{r, message=FALSE}
meta <- readr::read_tsv("input/sra_rnaseq2.tsv", na = "-")
orig_counts <- readr::read_tsv("output/feature_counts.tsv")
counts <- readr::read_tsv("output/feature_counts_deseq2_normed.tsv", na = "-")

counts <- dplyr::inner_join(
  meta,
  counts %>% pivot_longer(-id, names_to = "sra", values_to = "count"),
  by = "sra"
) %>%
dplyr::left_join(
  .,
  orig_counts %>% mutate(id = sub("gene-", "", Geneid)) %>% dplyr::select(id, gene_length = Length),
  by = "id"
) %>%
  dplyr::select(
    sra, bioproject,included,sample,
    organism,organism_tidyname,
    accession,accession_tidyname,
    host,host_tidyname,
    host_accession,host_accession_tidyname,
    tissue,hpi,plant_tissue,treatment,treatment_tidy,
    id,count,gene_length
  )

zero_counts <- counts %>% group_by(id) %>% summarise(count = sum(count)) %>% filter(count == 0.) %>% .[, "id"] %>% .[[1]]
counts <- counts %>% filter(!id %in% zero_counts)

rm(meta)
rm(orig_counts)

print(dim(counts))
head(counts)
```


There's a well known tendency for longer reads to have higher counts.
Normally this isn't a problem for DGE testing, because we're only comparing a gene with itself, so the length is constant.
However if we're looking for coexpression, we'll also be comparing the different gene lengths.
It may be fine if we use a metric that is unaffected by size (only correlation), like some of the rank methods.

So i'll have a look at out data to see if we need to normalise.

``` #{r, fig.height=10, fig.width=10}
ggplot(counts, aes(x = gene_length, y = count)) +
  geom_point(size = 1, alpha = 0.6) +
  facet_wrap(vars(bioproject))
```

There's a slight tendency for longer genes to have higher values.
The smaller genes make up the majority of genes with high expression, but this is really just because large genes are very rare.

``` #{r, fig.height=10, fig.width=10}
ggplot(counts, aes(x = gene_length, y = count)) +
  geom_point(size = 1, alpha = 0.6) +
  xlim(0, 2000) +
  facet_wrap(vars(bioproject))
```


So i'll divide the count by the gene length in kilobases.
This is roughly analogous to the common FPKM metric.

``` #{r, fig.height=10, fig.width=10}
ggplot(counts, aes(x = gene_length, y = count / (gene_length / 1000))) +
  geom_point(size = 1, alpha = 0.6) +
  facet_wrap(vars(bioproject))
```


``` #{r, fig.height=10, fig.width=10}
ggplot(counts, aes(x = gene_length, y = count / (gene_length / 1000))) +
  geom_point(size = 1, alpha = 0.6) +
  xlim(0, 2000) +
  facet_wrap(vars(bioproject))
```


I think actually that the unnormalised data looks ok.
And peaks we get at the lower values in the "normalised" counts worries me a bit.
It's probably fine either way, but I'll go with the unnormalised ones.


```{r}
gen_count_matrix <- function(df) {
  #zero_counts <- df %>% group_by(id) %>% summarise(count = sum(count)) %>% filter(count == 0.) %>% .[, "id"] %>% .[[1]]
  df <- df %>%
    pivot_wider(id_cols = "id", names_from = "sra", values_from = "count") %>%
    column_to_rownames("id") %>%
    as.matrix()
  
  return(df)
}

cor.test.p <- function(x, method = "pearson") {
  n <- (!is.na(x)) %*% t(!is.na(x))
  r <- cor(t(x), method = method, use = "pairwise.complete.obs")
  
  if (method %in% c("pearson", "spearman")) {
    t <- (r * sqrt(n - 2)) / sqrt(1 - r^2)
    p <- 2 * (1 - pt(abs(t), (n - 2)))
  } else if (method == "kendall") {
    p <- NULL
  } else {
    stop("Don't be silly darcy")
  }
  return(p)
}
```

```{r}
collapse_duplicates <- function(mat, precision = 4) {
  strings <- data.frame(
    id = rownames(mat),
    counts = do.call(paste, as.data.frame(round(mat, precision)))
  )

  strings <- strings %>%
    group_by(counts) %>%
    summarise(n = n_distinct(id), ids = paste0(unique(id), collapse = "__"), rep = first(id))

  strings <- strings[order(strings$ids), ]

  new_mat <- mat[strings[["rep"]], ]
  rownames(new_mat) <- strings[["ids"]]
  return(new_mat)
}
```

```{r}
collapse_approx_duplicates <- function(mat, method = "pearson", precision = (1 - 1e-5)) {
  cor_mat <- cor(t(mat), method = method) >= precision
  
  adj <- graph_from_adjacency_matrix(cor_mat, mode = "undirected", diag = TRUE)
  tmp <- components(adj)$membership

  gvar <- apply(mat, MARGIN = 1, var, na.rm = TRUE)
  gvar[is.na(gvar)] <- 0
  
  tmp <- data.frame(id = names(tmp), component = unname(tmp))
  gvar <- data.frame(id = names(gvar), var = unname(gvar))
  tmp <- left_join(
    tmp,
    gvar,
    by = "id"
  )

  tmp <- tmp[order(tmp$var, decreasing = TRUE, na.last = TRUE), ] %>%
    group_by(component) %>%
    summarise(rep = first(id), id = paste(id, collapse = "__", sep = "__"))
  
  new_mat <- mat[tmp[["rep"]], ]
  rownames(new_mat) <- tmp[["id"]]
  return(new_mat)
}
```

```{r, warning=FALSE}
wcounts <- gen_count_matrix(counts)
wcounts <- collapse_duplicates(wcounts)

print(sprintf("nrow before %d", nrow(wcounts)))
wcounts <- wcounts[rowSums(wcounts, na.rm = TRUE) > 0, ]
print(sprintf("nrow after %d", nrow(wcounts)))
wcounts[1:5, 1:5]
```

```{r}
wcounts_pca <- prcomp(wcounts)
ncomp <- max(which(cumsum(wcounts_pca$sdev / sum(wcounts_pca$sdev)) < 0.95))
ncomp
```


```{r}
wcounts_ica <- ica(wcounts, ncomp, maxit=1000)
wcounts_ica_s <- wcounts_ica$M
rownames(wcounts_ica_s) <- colnames(wcounts)
colnames(wcounts_ica_s) <- paste0("IC", seq_len(ncol(wcounts_ica_s)))
```

```{r}
ics <- right_join(
  counts %>% select(sra, bioproject, host = host_tidyname, host_accession = host_accession_tidyname, hpi, plant_tissue, sample) %>% unique(),
  as.data.frame(wcounts_ica_s) %>% rownames_to_column("sra"),
  by = "sra"
)
```



```{r, fig.width = 20, fig.height=10}
ics <- ics[order(ics$host, ics$host_accession, ics$sample), ]


ra <- rowAnnotation(
  bioproject = ics[, "bioproject"] %>% unlist(),
  host = ics[, "host"] %>% unlist(),
  acc = ics[, "host_accession"] %>% unlist(),
  hpi = ics[, "hpi"] %>% unlist()
)

Heatmap(
  matrix = ics %>% select(-bioproject, -host, -host_accession, -hpi, -plant_tissue, -sample) %>% column_to_rownames("sra") %>% as.matrix,
  clustering_distance_rows = "spearman",
  cluster_rows = FALSE,
  clustering_distance_columns = "spearman",
  right_annotation = ra,
  show_row_names = FALSE,
  show_column_names = FALSE,
  show_column_dend = FALSE
)
```

```{r}
tmp <- ics %>% select(-bioproject, -host, -host_accession, -hpi, -plant_tissue, -sample) %>% column_to_rownames("sra") %>% as.matrix
mask <- !is.na(ics$hpi)
tmp <- tmp[mask, ]
hpi <- ics$hpi[mask]

sort(apply(tmp, MARGIN = 2, FUN = function(x) {cor(x, hpi, method = "spearman")}))
```

```{r, fig.height=30, fig.width=40}
tmp <- ics %>% pivot_longer(cols = -c(sra, bioproject, host, host_accession, hpi, plant_tissue, sample), names_to = "IC")
gg <- tmp %>%  #group_by(bioproject, IC) %>% summarise(median = median(value), iqr = IQR(value)) %>%
  ggplot(aes(x = bioproject, y = value, colour = bioproject, shape = bioproject)) +
  geom_jitter()

gg <- gg_shape(gg)
gg + facet_wrap(vars(IC), scales = "free_y")
```

```{r, fig.height=30, fig.width=40}
tmp <- ics %>% pivot_longer(cols = -c(sra, bioproject, host, host_accession, hpi, plant_tissue, sample), names_to = "IC")
gg <- tmp %>%  #group_by(bioproject, IC) %>% summarise(median = median(value), iqr = IQR(value)) %>%
  ggplot(aes(x = host, y = value, colour = host, shape = host)) +
  geom_jitter()

gg <- gg_shape(gg)
gg + facet_wrap(vars(IC), scales = "free_y")
```


```{r}
gg <- ggplot(ics, aes(x = IC1, y = IC26, colour = host, shape = bioproject)) + geom_point()
gg <- gg_shape(gg)
gg
```




